{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Author:__ Bram Van de Sande\n",
    "\n",
    "__Date:__ 5 MAR 2018\n",
    "\n",
    "__Outline:__ Purpose is to create a collection of single-cell transcriptomics datasets to apply pySCENIC to with the attempt to create meta-regulomes, i.e. precalculated regulomes from a large databases.\n",
    "\n",
    "The idea would be to use these precalculated meta-regulomes to map new sc-experiment data to meta-regulome feature space and use this new feature space to find similar experiments in the database.\n",
    "\n",
    "Criteria for selecting datasets:\n",
    "- Data should be available as an expression matrix in TPM, FPKM and/or a log transformation (e.g. log2(TPM+1)). We will not map reads from FASTQ files ourselves.\n",
    "- Data should be related to human disease and/or clearly defined cell types.\n",
    "- Preferrably hg19. Which is the case for most datasets. \n",
    "- Check gene annotations and version used.\n",
    "\n",
    "__TODO:__\n",
    "- Check how many cell lines are true tumor cell lines (versus host cells)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import math\n",
    "from pyscenic.rnkdb import FeatherRankingDatabase as RankingDatabase\n",
    "from functools import partial\n",
    "import glob\n",
    "import mygene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATABASE_FOLDER = \"/Users/bramvandesande/Projects/lcb/databases/\"\n",
    "DATABASE_FNAME_500bp = os.path.join(DATABASE_FOLDER, \"hg19-500bp-upstream-10species.mc9nr.feather\")\n",
    "RESOURCES_FOLDER = '/Users/bramvandesande/Projects/lcb/resources/sc-experiments/'\n",
    "RESULTS_FOLDER = '/Users/bramvandesande/Projects/lcb/resources/sc-experiments/clean/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## List of human transcription factors\n",
    "\n",
    "Taken from http://www.cell.com/cell/fulltext/S0092-8674(18)30106-5 . See Supplementary Files.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GSE60361_C1-3005-Expression.txt      motifs-v8-nr.hgnc-m0.001-o0.0.tbl\n",
      "c6.all.v6.1.symbols.gmt.txt          motifs-v8-nr.mgi-m0.000-o0.0.tbl\n",
      "ensg2hgnc.tsv                        motifs-v8-nr.mgi-m0.001-o0.0.tbl\n",
      "mm_tfs.txt                           motifs-v8-nr.zfin-m0.000-o0.0.tbl\n",
      "modules_zeisel_2015.yaml             motifs-v8-nr.zfin-m0.001-o0.0.tbl\n",
      "motifs-v8-nr.fbgn-m0.000-o0.0.tbl    motifs-v9-nr.hgnc-m0.001-o0.0.tbl\n",
      "motifs-v8-nr.fbgn-m0.001-o0.0.tbl    motifs-v9-nr.mgi-m0.001-o0.0.tbl\n",
      "motifs-v8-nr.flybase-m0.000-o0.0.tbl regulomes_zeisel_2015.csv\n",
      "motifs-v8-nr.flybase-m0.001-o0.0.tbl \u001b[1m\u001b[36mrscenic\u001b[m\u001b[m\n",
      "motifs-v8-nr.hgnc-m0.000-o0.0.tbl    \u001b[1m\u001b[36msc-experiments\u001b[m\u001b[m\n"
     ]
    }
   ],
   "source": [
    "! ls ~/Projects/lcb/resources"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_excel(os.path.join(RESOURCES_FOLDER, 'human_tfs.xlsx'), sheet_name='Table S1. Related to Figure 1B',\n",
    "                  header=[0], skiprows=[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'Name', 'DBD', 'Unnamed: 3', 'TF assessment', 'Binding mode',\n",
       "       'Motif status', 'Notes', 'Comments', 'Committee notes', 'MTW Notes',\n",
       "       'TRH Notes', 'SL notes', 'AJ notes', 'Disagree on Assessment ',\n",
       "       'Disagree on Binding', 'Author1', 'Assesment1', 'Binding1', 'Comment1',\n",
       "       'Notes1', 'Author2', 'Assesment2', 'Binding2', 'Comment2', 'Notes2',\n",
       "       'Vaquerizas 2009 TF classification', 'CisBP considers it as a TF?',\n",
       "       'TFclass considers it as a TF? ', 'TF-CAT classification', 'Is a GO TF',\n",
       "       'PDB'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "human_tfs = list(df[df['Unnamed: 3'] == 'Yes'].Name.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1639"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(human_tfs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(RESULTS_FOLDER, 'tfs.txt'), 'w') as file:\n",
    "    file.write('\\n'.join(human_tfs) + '\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Source I: SCPortalen (http://single-cell.clst.riken.jp/)\n",
    "\n",
    "Total list of FPKM expression matrics can be found at http://single-cell.clst.riken.jp/non_riken_data/bd_fpkm_file_list.php?pagesize=-1 .\n",
    "\n",
    "Might need to convert FPKM to TPM units (https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/). Also, need to convert ENSEMBL gene identifiers (ENSG) to HGNC nomenclature. Cluster assignment might not be available.\n",
    "\n",
    "| ID | Disease/Cell type | Download URL |\n",
    "|----|--------------------------|--------------|\n",
    "| DRA001287 | Lung adenocarcinoma | http://single-cell.clst.riken.jp/fpkm_tables/DRA001287_fpkm.table.gz |\n",
    "| DRA002730 | Lung adenocarcinoma | http://single-cell.clst.riken.jp/fpkm_tables/DRA002730_fpkm.table.gz |\n",
    "| E-MTAB-5061 | Human pancreas (healthy and diabetes type II) | http://single-cell.clst.riken.jp/fpkm_tables/E-MTAB-5061_fpkm.table.gz | \n",
    "| GSE57872 | Glioblastoma | http://single-cell.clst.riken.jp/fpkm_tables/GSE57872_fpkm.table.gz |\n",
    "| GSE73121 | Renal Cell Carcinoma | http://single-cell.clst.riken.jp/fpkm_tables/GSE73121_fpkm.table.gz | "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Source II: Single Cell Portal (https://portals.broadinstitute.org/single_cell)\n",
    "\n",
    "Gene nomenclature is already HGNC.\n",
    "\n",
    "| Disease/Cell type | Number of cells | Number of clusters | Publication | Unit | URL | \n",
    "|---------|-----------------|--------------------|-------------|------|-----|\n",
    "| Astrocytoma | 6341 | 4 clusters | doi: 10.1126/science.aai8478 | log2(TPM/10+1), for all genes and cells | https://portals.broadinstitute.org/single_cell/study/single-cell-rna-seq-analysis-of-astrocytoma#study-summary |\n",
    "| Blood dendritic cells and monocytes | 1078 | 10 clusters | doi: 10.1126/science.aah4573 | Log transformed and filtered TPMs | https://portals.broadinstitute.org/single_cell/study/atlas-of-human-blood-dendritic-cells-and-monocytes#study-summary |\n",
    "| Melanoma | 4646 | 7 clusters | doi: 10.1126/science.aad0501 | log2(TPM/10+1) of all genes in all cells | https://portals.broadinstitute.org/single_cell/study/melanoma-intra-tumor-heterogeneity |\n",
    "| Glioblastoma | 430 | ? | doi: 10.1126/science.1254257 | log2(TPM+1) expression matrix for all cells profiled including 5948 genes with highest expression | https://portals.broadinstitute.org/single_cell/study/glioblastoma-intra-tumor-heterogeneity  |\n",
    "| Oligodendroglioma | 4347 | ? | doi:10.1038/nature20123 | log2(TPM/10+1) of all genes in all cells |  https://portals.broadinstitute.org/single_cell/study/oligodendroglioma-intra-tumor-heterogeneity#study-summary | "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Source III: GEO (https://www.ncbi.nlm.nih.gov/geo)\n",
    "\n",
    "Searched Geo Datasets (https://www.ncbi.nlm.nih.gov/gds/) using the following query:\n",
    "\n",
    "```\n",
    "\"homo sapiens\"[Organism] AND \"expression profiling by high throughput sequencing\"[DataSet Type] AND \"single cell\" AND \"Neoplasms\"[Mesh] \n",
    "```\n",
    "\n",
    "Selected for the large sample dataset + included dataset for prevalent tumor types (https://www.cancer.gov/types/common-cancers). Bladder and prostate cancer is missing.\n",
    "\n",
    "| Downloaded? | Source | Cancer subtype | # cells | Dataset ID | Publication ID | Unit | Method | URL |\n",
    "|------|--------|--------|---------|------------|----------------|------|--------|-----|\n",
    "| X (23686x6341) | BROAD (GEO) | IDH-mutant astrocytoma | 6341 | GSE89567 | DOI: 10.1126/science.aai8478 | Expression levels were quantified as Ei,j=log2(TPMi,j/10+1), where TPMi,j refers to transcriptper-million for gene i in sample j, as calculated by RSEM.  Tab-delimited text file containing the normalized expression levels (E) for 23,686 analyzed genes (rows, first, column indicates Gene Symbols) across 6341 astrocytoma cells (columns, sample names indicated at the top row). | SMART-seq2 | https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE89567 (https://portals.broadinstitute.org/single_cell/study/single-cell-rna-seq-analysis-of-astrocytoma#study-summary) |\n",
    "| X | GEO | Head and neck cancer | 5902 | GSE103322 | PMID: 29198524 | Tab-delimited text file containing TPM values ( transcript-per-million reads) for 23,686 analyzed genes (rows), across 5902 cells (columns, sample names indicated at the top row). | SMART-Seq2 | https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE103322 |\n",
    "| X | BROAD (GEO) | Oligodendroglioma | 4347 | GSE70630 | DOI: 10.1038/nature20123 | log2(TPM/10+1) of all genes in all cells | SMART-Seq2 | https://portals.broadinstitute.org/single_cell/study/oligodendroglioma-intra-tumor-heterogeneity#study-summary (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE70630) | \n",
    "| X (23686x4646)| BROAD (GEO) | Melanoma | 4646 | GSE72056 | DOI: 10.1126/science.aad0501 | log2(TPM/10+1) of all genes in all cells | SMART-Seq2 | https://portals.broadinstitute.org/single_cell/study/melanoma-intra-tumor-heterogeneity (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE72056)  |\n",
    "| X | BROAD (GEO +SCPortalen) | Glioblastoma | 430 | GSE57872 | DOI: 10.1126/science.1254257 | log2(TPM+1) expression matrix for all cells profiled including 5948 genes with highest expression. | SMART-seq | https://portals.broadinstitute.org/single_cell/study/glioblastoma-intra-tumor-heterogeneity (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE57872) |\n",
    "| - (Raw reads) | GEO | Glioblastoma | 3589 | GSE84465 | PMID: 29091775 | Tab-delimited text files with raw read values for each sample | SMART-Seq2 | https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE84465 |\n",
    "| X | GEO | Chronic Myeloid Leukemia | 2289 | GSE76312 | PMID: 28504724 | RPKM | Fluidigm | https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE76312 |\n",
    "| X | GEO | Colorectal tumors | 1591 | GSE81861 | DOI: 10.1038/ng.3818 | FPKM | Fluidigm | https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE81861 | \n",
    "| X | GEO | Lung adenocarcinoma | 126  | GSE69405 | DOI: 10.1186/s13059-015-0692-3 | TPM | SMART-Seq | https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE69405 |\n",
    "| X | GEO | Breastcancer | 549 | GSE75688 | PMID: 28474673 | A processed file is tab-delimited and contains the raw TPM values for 549 single cells and 14 bulk tumors. | integrated fluidic circuit (IFC) chip | https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE75688 |\n",
    "| - (HGNC gene identifiers) | GEO (SCPortalen) | RCC | 116 | GSE73121 | PMID: 27139883 | Each row of the tab-delimited text file includes ENSEMBL gene ID,ENSEMBL transcript ID, transcript length, effective_length, expected_count, TPM, FPKM for samples. | Fluidigm + SMARTer | https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE73121 |\n",
    "\n",
    "__Disclaimer:__ The current compendium suffers from the limited number of cells available for some very prevalent tumor types (e.g. lung, breast and colorectal).\n",
    "The reason for this is that single cell has been pioneered in brain research. Hence, the dominance of the brain tumors in this compendium.\n",
    "\n",
    "__Solution:__ For lung we can use the in-house dataset from Bernard Thienpont with over 55K cells taken from multiple patients."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preparation of datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GSE103322.mtx+clusters.txt\n",
      "GSE57872.mtx.txt\n",
      "GSE69405.mtx.txt\n",
      "GSE70630.clusters.txt\n",
      "GSE70630.mtx.txt\n",
      "GSE72056.clusters.txt\n",
      "GSE72056.mtx.txt\n",
      "\u001b[1m\u001b[36mGSE73121\u001b[m\u001b[m\n",
      "GSE75688.mtx.txt\n",
      "GSE76312.clusters.txt\n",
      "GSE76312.mtx.txt\n",
      "GSE81861_CRC_NM_all_cells_FPKM.csv\n",
      "GSE81861_CRC_NM_epithelial_cells_FPKM.csv\n",
      "GSE81861_CRC_tumor_all_cells_FPKM.csv\n",
      "GSE81861_CRC_tumor_epithelial_cells_FPKM.csv\n",
      "GSE81861_Cell_Line_FPKM.csv\n",
      "GSE89567.clusters.txt\n",
      "GSE89567.mtx.txt\n",
      "bernard_thienpont_2017_lungcancer.txt\n",
      "bernard_thienpont_2017_lungcancer_grnboost.RData\n",
      "bernard_thienpont_2017_lungcancer_grnboost.txt\n",
      "bernard_thienpont_2017_lungcancer_metadata.txt\n",
      "\u001b[1m\u001b[36mclean\u001b[m\u001b[m\n",
      "human_tfs.xlsx\n"
     ]
    }
   ],
   "source": [
    "!ls {RESOURCES_FOLDER}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_in_db = RankingDatabase(DATABASE_FNAME_500bp, name=\"500bp\", nomenclature=\"MGI\").geneset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### B. Thienpont - Lungcarcinoma\n",
    "\n",
    "GRNBoost data is already available. The disadvantage of using the already available data is that we might not have used the same list of TFs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "adjacencies = pd.read_csv(os.path.join(RESOURCES_FOLDER, 'bernard_thienpont_2017_lungcancer_grnboost.txt'), sep=',')\n",
    "del adjacencies['Unnamed: 0']\n",
    "adjacencies.columns = ['TF', 'target', 'importance']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF</th>\n",
       "      <th>target</th>\n",
       "      <th>importance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GATA2</td>\n",
       "      <td>LTC4S</td>\n",
       "      <td>0.877992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GATA2</td>\n",
       "      <td>CPA3</td>\n",
       "      <td>0.803889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GATA2</td>\n",
       "      <td>TPSAB1</td>\n",
       "      <td>0.679747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GATA2</td>\n",
       "      <td>HPGDS</td>\n",
       "      <td>0.653688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GATA2</td>\n",
       "      <td>MS4A2</td>\n",
       "      <td>0.651024</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      TF  target  importance\n",
       "0  GATA2   LTC4S    0.877992\n",
       "1  GATA2    CPA3    0.803889\n",
       "2  GATA2  TPSAB1    0.679747\n",
       "3  GATA2   HPGDS    0.653688\n",
       "4  GATA2   MS4A2    0.651024"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adjacencies.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "adjacencies.to_csv(os.path.join(RESULTS_FOLDER, 'b_thienpont.net.csv'), sep=',', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GSE89567 - Astrocytoma"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_gse89567(fname, genes_in_db):\n",
    "    # Load CSV file\n",
    "    mtx = pd.read_csv(os.path.join(RESOURCES_FOLDER, fname), sep='\\t', index_col=0)\n",
    "    \n",
    "    # Remove duplicate gene symbols (keep last when sorted ascending order according to row sum)\n",
    "    mtx = mtx.iloc[mtx.sum(axis=1).argsort()]\n",
    "    mtx = mtx[~mtx.index.duplicated(keep='last')]\n",
    "    \n",
    "    # Keep only genes that are in the ranking database\n",
    "    mtx = mtx[mtx.index.isin(genes_in_db)]\n",
    "    \n",
    "    return mtx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx = process_gse89567('GSE89567.mtx.txt', genes_in_db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MGH42_P7_A01</th>\n",
       "      <th>MGH42_P7_A02</th>\n",
       "      <th>MGH42_P7_A03</th>\n",
       "      <th>MGH42_P7_A04</th>\n",
       "      <th>MGH42_P7_A05</th>\n",
       "      <th>MGH42_P7_A07</th>\n",
       "      <th>MGH42_P7_A09</th>\n",
       "      <th>MGH42_P7_A11</th>\n",
       "      <th>MGH42_P7_A12</th>\n",
       "      <th>MGH42_P7_B02</th>\n",
       "      <th>...</th>\n",
       "      <th>MGH107neg_P2_E06</th>\n",
       "      <th>MGH107pos_P2_B03</th>\n",
       "      <th>MGH107neg_P1_F03</th>\n",
       "      <th>MGH107neg_P1_G06</th>\n",
       "      <th>MGH107neg_P2_H03</th>\n",
       "      <th>MGH107neg_P2_C05</th>\n",
       "      <th>MGH107pos_P2_D07</th>\n",
       "      <th>MGH107neg_P1_E01</th>\n",
       "      <th>MGH107pos_P2_G09</th>\n",
       "      <th>MGH107neg_P1_D06</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR760</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR202</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR200C</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR200B</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR200A</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 6341 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         MGH42_P7_A01  MGH42_P7_A02  MGH42_P7_A03  MGH42_P7_A04  MGH42_P7_A05  \\\n",
       "Gene                                                                            \n",
       "MIR760            0.0           0.0           0.0           0.0           0.0   \n",
       "MIR202            0.0           0.0           0.0           0.0           0.0   \n",
       "MIR200C           0.0           0.0           0.0           0.0           0.0   \n",
       "MIR200B           0.0           0.0           0.0           0.0           0.0   \n",
       "MIR200A           0.0           0.0           0.0           0.0           0.0   \n",
       "\n",
       "         MGH42_P7_A07  MGH42_P7_A09  MGH42_P7_A11  MGH42_P7_A12  MGH42_P7_B02  \\\n",
       "Gene                                                                            \n",
       "MIR760            0.0           0.0           0.0           0.0           0.0   \n",
       "MIR202            0.0           0.0           0.0           0.0           0.0   \n",
       "MIR200C           0.0           0.0           0.0           0.0           0.0   \n",
       "MIR200B           0.0           0.0           0.0           0.0           0.0   \n",
       "MIR200A           0.0           0.0           0.0           0.0           0.0   \n",
       "\n",
       "               ...         MGH107neg_P2_E06  MGH107pos_P2_B03  \\\n",
       "Gene           ...                                              \n",
       "MIR760         ...                      0.0               0.0   \n",
       "MIR202         ...                      0.0               0.0   \n",
       "MIR200C        ...                      0.0               0.0   \n",
       "MIR200B        ...                      0.0               0.0   \n",
       "MIR200A        ...                      0.0               0.0   \n",
       "\n",
       "         MGH107neg_P1_F03  MGH107neg_P1_G06  MGH107neg_P2_H03  \\\n",
       "Gene                                                            \n",
       "MIR760                0.0               0.0               0.0   \n",
       "MIR202                0.0               0.0               0.0   \n",
       "MIR200C               0.0               0.0               0.0   \n",
       "MIR200B               0.0               0.0               0.0   \n",
       "MIR200A               0.0               0.0               0.0   \n",
       "\n",
       "         MGH107neg_P2_C05  MGH107pos_P2_D07  MGH107neg_P1_E01  \\\n",
       "Gene                                                            \n",
       "MIR760                0.0               0.0               0.0   \n",
       "MIR202                0.0               0.0               0.0   \n",
       "MIR200C               0.0               0.0               0.0   \n",
       "MIR200B               0.0               0.0               0.0   \n",
       "MIR200A               0.0               0.0               0.0   \n",
       "\n",
       "         MGH107pos_P2_G09  MGH107neg_P1_D06  \n",
       "Gene                                         \n",
       "MIR760                0.0               0.0  \n",
       "MIR202                0.0               0.0  \n",
       "MIR200C               0.0               0.0  \n",
       "MIR200B               0.0               0.0  \n",
       "MIR200A               0.0               0.0  \n",
       "\n",
       "[5 rows x 6341 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mtx.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20684, 6341)"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mtx.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1567"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(mtx.index).intersection(set(human_tfs)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx.to_csv(os.path.join(RESULTS_FOLDER, 'GSE89567.mtx.tsv'), sep='\\t', index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GSE103322 - Head and neck cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_gse103322(fname, genes_in_db):\n",
    "    # Load CSV file\n",
    "    mtx = pd.read_csv(os.path.join(RESOURCES_FOLDER, fname), sep='\\t', index_col=0, skiprows=[1,2,3,4,5])\n",
    "    \n",
    "    # Extract gene symbol\n",
    "    mtx.index = list(map(lambda g: g[1:-1], mtx.index))\n",
    "    \n",
    "    # Remove duplicate gene symbols (keep last when sorted ascending order according to row sum)\n",
    "    mtx = mtx.iloc[mtx.sum(axis=1).argsort()]\n",
    "    mtx = mtx[~mtx.index.duplicated(keep='last')]\n",
    "    \n",
    "    # From TPM to log2(TPM/10+1)\n",
    "    mtx = mtx.applymap(lambda tpm: math.log2((tpm/10.0) + 1.0))\n",
    "    \n",
    "    # Keep only genes that are in the ranking database\n",
    "    mtx = mtx[mtx.index.isin(genes_in_db)]\n",
    "    \n",
    "    return mtx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx = process_gse103322('GSE103322.mtx+clusters.txt', genes_in_db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HN28_P15_D06_S330_comb</th>\n",
       "      <th>HN28_P6_G05_S173_comb</th>\n",
       "      <th>HN26_P14_D11_S239_comb</th>\n",
       "      <th>HN26_P14_H05_S281_comb</th>\n",
       "      <th>HN26_P25_H09_S189_comb</th>\n",
       "      <th>HN26_P14_H06_S282_comb</th>\n",
       "      <th>HN25_P25_C04_S316_comb</th>\n",
       "      <th>HN26_P25_A11_S107_comb</th>\n",
       "      <th>HN26_P25_C09_S129_comb</th>\n",
       "      <th>HNSCC26_P24_H05_S377_comb</th>\n",
       "      <th>...</th>\n",
       "      <th>HNSCC20_P3_B10_S22_comb</th>\n",
       "      <th>HNSCC20_P13_B11_S215_comb</th>\n",
       "      <th>HNSCC20_P3_C08_S32_comb</th>\n",
       "      <th>HNSCC17_P4_H03_S183_comb</th>\n",
       "      <th>HNSCC20_P3_F09_S69_comb</th>\n",
       "      <th>HNSCC17_P4_G12_S180_comb</th>\n",
       "      <th>HNSCC20_P13_C05_S221_comb</th>\n",
       "      <th>HNSCC17_P4_C12_S132_comb</th>\n",
       "      <th>HNSCC20_P3_H08_S92_comb</th>\n",
       "      <th>HNSCC20_P3_G06_S78_comb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SNORD113-9</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAGEB16</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNORA49</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR26A1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR485</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 5902 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            HN28_P15_D06_S330_comb  HN28_P6_G05_S173_comb  \\\n",
       "SNORD113-9                     0.0                    0.0   \n",
       "MAGEB16                        0.0                    0.0   \n",
       "SNORA49                        0.0                    0.0   \n",
       "MIR26A1                        0.0                    0.0   \n",
       "MIR485                         0.0                    0.0   \n",
       "\n",
       "            HN26_P14_D11_S239_comb  HN26_P14_H05_S281_comb  \\\n",
       "SNORD113-9                     0.0                     0.0   \n",
       "MAGEB16                        0.0                     0.0   \n",
       "SNORA49                        0.0                     0.0   \n",
       "MIR26A1                        0.0                     0.0   \n",
       "MIR485                         0.0                     0.0   \n",
       "\n",
       "            HN26_P25_H09_S189_comb  HN26_P14_H06_S282_comb  \\\n",
       "SNORD113-9                     0.0                     0.0   \n",
       "MAGEB16                        0.0                     0.0   \n",
       "SNORA49                        0.0                     0.0   \n",
       "MIR26A1                        0.0                     0.0   \n",
       "MIR485                         0.0                     0.0   \n",
       "\n",
       "            HN25_P25_C04_S316_comb  HN26_P25_A11_S107_comb  \\\n",
       "SNORD113-9                     0.0                     0.0   \n",
       "MAGEB16                        0.0                     0.0   \n",
       "SNORA49                        0.0                     0.0   \n",
       "MIR26A1                        0.0                     0.0   \n",
       "MIR485                         0.0                     0.0   \n",
       "\n",
       "            HN26_P25_C09_S129_comb  HNSCC26_P24_H05_S377_comb  \\\n",
       "SNORD113-9                     0.0                        0.0   \n",
       "MAGEB16                        0.0                        0.0   \n",
       "SNORA49                        0.0                        0.0   \n",
       "MIR26A1                        0.0                        0.0   \n",
       "MIR485                         0.0                        0.0   \n",
       "\n",
       "                     ...             HNSCC20_P3_B10_S22_comb  \\\n",
       "SNORD113-9           ...                                 0.0   \n",
       "MAGEB16              ...                                 0.0   \n",
       "SNORA49              ...                                 0.0   \n",
       "MIR26A1              ...                                 0.0   \n",
       "MIR485               ...                                 0.0   \n",
       "\n",
       "            HNSCC20_P13_B11_S215_comb  HNSCC20_P3_C08_S32_comb  \\\n",
       "SNORD113-9                        0.0                      0.0   \n",
       "MAGEB16                           0.0                      0.0   \n",
       "SNORA49                           0.0                      0.0   \n",
       "MIR26A1                           0.0                      0.0   \n",
       "MIR485                            0.0                      0.0   \n",
       "\n",
       "            HNSCC17_P4_H03_S183_comb  HNSCC20_P3_F09_S69_comb  \\\n",
       "SNORD113-9                       0.0                      0.0   \n",
       "MAGEB16                          0.0                      0.0   \n",
       "SNORA49                          0.0                      0.0   \n",
       "MIR26A1                          0.0                      0.0   \n",
       "MIR485                           0.0                      0.0   \n",
       "\n",
       "            HNSCC17_P4_G12_S180_comb  HNSCC20_P13_C05_S221_comb  \\\n",
       "SNORD113-9                       0.0                        0.0   \n",
       "MAGEB16                          0.0                        0.0   \n",
       "SNORA49                          0.0                        0.0   \n",
       "MIR26A1                          0.0                        0.0   \n",
       "MIR485                           0.0                        0.0   \n",
       "\n",
       "            HNSCC17_P4_C12_S132_comb  HNSCC20_P3_H08_S92_comb  \\\n",
       "SNORD113-9                       0.0                      0.0   \n",
       "MAGEB16                          0.0                      0.0   \n",
       "SNORA49                          0.0                      0.0   \n",
       "MIR26A1                          0.0                      0.0   \n",
       "MIR485                           0.0                      0.0   \n",
       "\n",
       "            HNSCC20_P3_G06_S78_comb  \n",
       "SNORD113-9                      0.0  \n",
       "MAGEB16                         0.0  \n",
       "SNORA49                         0.0  \n",
       "MIR26A1                         0.0  \n",
       "MIR485                          0.0  \n",
       "\n",
       "[5 rows x 5902 columns]"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mtx.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20684, 5902)"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mtx.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1567"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(mtx.index).intersection(set(human_tfs)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx.to_csv(os.path.join(RESULTS_FOLDER, 'GSE103322.mtx.tsv'), sep='\\t', index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GSE70630 - Oligodendroglioma"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_gse70630(fname, genes_in_db):\n",
    "    # Load CSV file\n",
    "    mtx = pd.read_csv(os.path.join(RESOURCES_FOLDER, fname), sep='\\t', index_col=0)\n",
    "    \n",
    "    # Remove duplicate gene symbols (keep last when sorted ascending order according to row sum)\n",
    "    mtx = mtx.iloc[mtx.sum(axis=1).argsort()]\n",
    "    mtx = mtx[~mtx.index.duplicated(keep='last')]\n",
    "    \n",
    "    # Keep only genes that are in the ranking database\n",
    "    mtx = mtx[mtx.index.isin(genes_in_db)]\n",
    "    \n",
    "    return mtx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx = process_gse70630('GSE70630.mtx.txt', genes_in_db)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove NaNs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "N_nans = mtx.isnull().values.sum()\n",
    "from operator import mul\n",
    "from functools import reduce\n",
    "N_totals = reduce(mul, mtx.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.018560091878683019"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "N_nans/N_totals*100.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx = mtx.fillna(value=0.0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MGH36_P6_A12</th>\n",
       "      <th>MGH36_P6_H09</th>\n",
       "      <th>MGH53_P4_G04</th>\n",
       "      <th>MGH36_P10_G12</th>\n",
       "      <th>MGH53_P2_H12</th>\n",
       "      <th>MGH53_P4_D10</th>\n",
       "      <th>MGH53_P4_D01</th>\n",
       "      <th>MGH36_P6_B07</th>\n",
       "      <th>MGH36_P10_B12</th>\n",
       "      <th>MGH53_P2_G11</th>\n",
       "      <th>...</th>\n",
       "      <th>93_P10_H06</th>\n",
       "      <th>93_P8_B12</th>\n",
       "      <th>93_P8_D09</th>\n",
       "      <th>93_P9_D11</th>\n",
       "      <th>93_P10_G08</th>\n",
       "      <th>93_P8_H06</th>\n",
       "      <th>93_P9_C07</th>\n",
       "      <th>93_P8_A12</th>\n",
       "      <th>93_P8_C01</th>\n",
       "      <th>93_P9_F06</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENE</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR3202-2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR520G</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR520F</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR520E</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR520D</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 4347 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           MGH36_P6_A12  MGH36_P6_H09  MGH53_P4_G04  MGH36_P10_G12  \\\n",
       "GENE                                                                 \n",
       "MIR3202-2           0.0           0.0           0.0            0.0   \n",
       "MIR520G             0.0           0.0           0.0            0.0   \n",
       "MIR520F             0.0           0.0           0.0            0.0   \n",
       "MIR520E             0.0           0.0           0.0            0.0   \n",
       "MIR520D             0.0           0.0           0.0            0.0   \n",
       "\n",
       "           MGH53_P2_H12  MGH53_P4_D10  MGH53_P4_D01  MGH36_P6_B07  \\\n",
       "GENE                                                                \n",
       "MIR3202-2           0.0           0.0           0.0           0.0   \n",
       "MIR520G             0.0           0.0           0.0           0.0   \n",
       "MIR520F             0.0           0.0           0.0           0.0   \n",
       "MIR520E             0.0           0.0           0.0           0.0   \n",
       "MIR520D             0.0           0.0           0.0           0.0   \n",
       "\n",
       "           MGH36_P10_B12  MGH53_P2_G11    ...      93_P10_H06  93_P8_B12  \\\n",
       "GENE                                      ...                              \n",
       "MIR3202-2            0.0           0.0    ...             0.0        0.0   \n",
       "MIR520G              0.0           0.0    ...             0.0        0.0   \n",
       "MIR520F              0.0           0.0    ...             0.0        0.0   \n",
       "MIR520E              0.0           0.0    ...             0.0        0.0   \n",
       "MIR520D              0.0           0.0    ...             0.0        0.0   \n",
       "\n",
       "           93_P8_D09  93_P9_D11  93_P10_G08  93_P8_H06  93_P9_C07  93_P8_A12  \\\n",
       "GENE                                                                           \n",
       "MIR3202-2        0.0        0.0         0.0        0.0        0.0        0.0   \n",
       "MIR520G          0.0        0.0         0.0        0.0        0.0        0.0   \n",
       "MIR520F          0.0        0.0         0.0        0.0        0.0        0.0   \n",
       "MIR520E          0.0        0.0         0.0        0.0        0.0        0.0   \n",
       "MIR520D          0.0        0.0         0.0        0.0        0.0        0.0   \n",
       "\n",
       "           93_P8_C01  93_P9_F06  \n",
       "GENE                             \n",
       "MIR3202-2        0.0        0.0  \n",
       "MIR520G          0.0        0.0  \n",
       "MIR520F          0.0        0.0  \n",
       "MIR520E          0.0        0.0  \n",
       "MIR520D          0.0        0.0  \n",
       "\n",
       "[5 rows x 4347 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mtx.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20684, 4347)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mtx.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1567"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(mtx.index).intersection(set(human_tfs)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx.to_csv(os.path.join(RESULTS_FOLDER, 'GSE70630.mtx.tsv'), sep='\\t', index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GSE72056 - Melanoma"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_gse72056(fname, genes_in_db):\n",
    "    # Load CSV file\n",
    "    mtx = pd.read_csv(os.path.join(RESOURCES_FOLDER, fname), sep='\\t', index_col=0)\n",
    "    \n",
    "    # Remove duplicate gene symbols (keep last when sorted ascending order according to row sum)\n",
    "    mtx = mtx.iloc[mtx.sum(axis=1).argsort()]\n",
    "    mtx = mtx[~mtx.index.duplicated(keep='last')]\n",
    "    \n",
    "    # Keep only genes that are in the ranking database\n",
    "    mtx = mtx[mtx.index.isin(genes_in_db)]\n",
    "    \n",
    "    return mtx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx = process_gse72056('GSE72056.mtx.txt', genes_in_db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cy72_CD45_H02_S758_comb</th>\n",
       "      <th>CY58_1_CD45_B02_S974_comb</th>\n",
       "      <th>Cy71_CD45_D08_S524_comb</th>\n",
       "      <th>Cy81_FNA_CD45_B01_S301_comb</th>\n",
       "      <th>Cy80_II_CD45_B07_S883_comb</th>\n",
       "      <th>Cy81_Bulk_CD45_B10_S118_comb</th>\n",
       "      <th>Cy72_CD45_D09_S717_comb</th>\n",
       "      <th>Cy74_CD45_A03_S387_comb</th>\n",
       "      <th>Cy71_CD45_B05_S497_comb</th>\n",
       "      <th>Cy80_II_CD45_C09_S897_comb</th>\n",
       "      <th>...</th>\n",
       "      <th>CY75_1_CD45_CD8_7__S265_comb</th>\n",
       "      <th>CY75_1_CD45_CD8_3__S127_comb</th>\n",
       "      <th>CY75_1_CD45_CD8_1__S61_comb</th>\n",
       "      <th>CY75_1_CD45_CD8_1__S12_comb</th>\n",
       "      <th>CY75_1_CD45_CD8_1__S25_comb</th>\n",
       "      <th>CY75_1_CD45_CD8_7__S223_comb</th>\n",
       "      <th>CY75_1_CD45_CD8_1__S65_comb</th>\n",
       "      <th>CY75_1_CD45_CD8_1__S93_comb</th>\n",
       "      <th>CY75_1_CD45_CD8_1__S76_comb</th>\n",
       "      <th>CY75_1_CD45_CD8_7__S274_comb</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENE</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR3924</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR3910-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR182</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR3665</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR3179-3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 4645 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Cy72_CD45_H02_S758_comb  CY58_1_CD45_B02_S974_comb  \\\n",
       "GENE                                                            \n",
       "MIR3924                        0.0                        0.0   \n",
       "MIR3910-1                      0.0                        0.0   \n",
       "MIR182                         0.0                        0.0   \n",
       "MIR3665                        0.0                        0.0   \n",
       "MIR3179-3                      0.0                        0.0   \n",
       "\n",
       "           Cy71_CD45_D08_S524_comb  Cy81_FNA_CD45_B01_S301_comb  \\\n",
       "GENE                                                              \n",
       "MIR3924                        0.0                          0.0   \n",
       "MIR3910-1                      0.0                          0.0   \n",
       "MIR182                         0.0                          0.0   \n",
       "MIR3665                        0.0                          0.0   \n",
       "MIR3179-3                      0.0                          0.0   \n",
       "\n",
       "           Cy80_II_CD45_B07_S883_comb  Cy81_Bulk_CD45_B10_S118_comb  \\\n",
       "GENE                                                                  \n",
       "MIR3924                           0.0                           0.0   \n",
       "MIR3910-1                         0.0                           0.0   \n",
       "MIR182                            0.0                           0.0   \n",
       "MIR3665                           0.0                           0.0   \n",
       "MIR3179-3                         0.0                           0.0   \n",
       "\n",
       "           Cy72_CD45_D09_S717_comb  Cy74_CD45_A03_S387_comb  \\\n",
       "GENE                                                          \n",
       "MIR3924                        0.0                      0.0   \n",
       "MIR3910-1                      0.0                      0.0   \n",
       "MIR182                         0.0                      0.0   \n",
       "MIR3665                        0.0                      0.0   \n",
       "MIR3179-3                      0.0                      0.0   \n",
       "\n",
       "           Cy71_CD45_B05_S497_comb  Cy80_II_CD45_C09_S897_comb  \\\n",
       "GENE                                                             \n",
       "MIR3924                        0.0                         0.0   \n",
       "MIR3910-1                      0.0                         0.0   \n",
       "MIR182                         0.0                         0.0   \n",
       "MIR3665                        0.0                         0.0   \n",
       "MIR3179-3                      0.0                         0.0   \n",
       "\n",
       "                       ...               CY75_1_CD45_CD8_7__S265_comb  \\\n",
       "GENE                   ...                                              \n",
       "MIR3924                ...                                        0.0   \n",
       "MIR3910-1              ...                                        0.0   \n",
       "MIR182                 ...                                        0.0   \n",
       "MIR3665                ...                                        0.0   \n",
       "MIR3179-3              ...                                        0.0   \n",
       "\n",
       "           CY75_1_CD45_CD8_3__S127_comb  CY75_1_CD45_CD8_1__S61_comb  \\\n",
       "GENE                                                                   \n",
       "MIR3924                             0.0                          0.0   \n",
       "MIR3910-1                           0.0                          0.0   \n",
       "MIR182                              0.0                          0.0   \n",
       "MIR3665                             0.0                          0.0   \n",
       "MIR3179-3                           0.0                          0.0   \n",
       "\n",
       "           CY75_1_CD45_CD8_1__S12_comb  CY75_1_CD45_CD8_1__S25_comb  \\\n",
       "GENE                                                                  \n",
       "MIR3924                            0.0                          0.0   \n",
       "MIR3910-1                          0.0                          0.0   \n",
       "MIR182                             0.0                          0.0   \n",
       "MIR3665                            0.0                          0.0   \n",
       "MIR3179-3                          0.0                          0.0   \n",
       "\n",
       "           CY75_1_CD45_CD8_7__S223_comb  CY75_1_CD45_CD8_1__S65_comb  \\\n",
       "GENE                                                                   \n",
       "MIR3924                             0.0                          0.0   \n",
       "MIR3910-1                           0.0                          0.0   \n",
       "MIR182                              0.0                          0.0   \n",
       "MIR3665                             0.0                          0.0   \n",
       "MIR3179-3                           0.0                          0.0   \n",
       "\n",
       "           CY75_1_CD45_CD8_1__S93_comb  CY75_1_CD45_CD8_1__S76_comb  \\\n",
       "GENE                                                                  \n",
       "MIR3924                            0.0                          0.0   \n",
       "MIR3910-1                          0.0                          0.0   \n",
       "MIR182                             0.0                          0.0   \n",
       "MIR3665                            0.0                          0.0   \n",
       "MIR3179-3                          0.0                          0.0   \n",
       "\n",
       "           CY75_1_CD45_CD8_7__S274_comb  \n",
       "GENE                                     \n",
       "MIR3924                             0.0  \n",
       "MIR3910-1                           0.0  \n",
       "MIR182                              0.0  \n",
       "MIR3665                             0.0  \n",
       "MIR3179-3                           0.0  \n",
       "\n",
       "[5 rows x 4645 columns]"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mtx.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20682, 4645)"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mtx.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1567"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(mtx.index).intersection(set(human_tfs)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx.to_csv(os.path.join(RESULTS_FOLDER, 'GSE72056.mtx.tsv'), sep='\\t', index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GSE57872 - Glioblastoma"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_gse57872(fname, genes_in_db):\n",
    "    # Load CSV file\n",
    "    mtx = pd.read_csv(os.path.join(RESOURCES_FOLDER, fname), sep='\\t', index_col=0)\n",
    "    \n",
    "    # Remove duplicate gene symbols (keep last when sorted ascending order according to row sum)\n",
    "    mtx = mtx.iloc[mtx.sum(axis=1).argsort()]\n",
    "    mtx = mtx[~mtx.index.duplicated(keep='last')]\n",
    "    \n",
    "    # Keep only genes that are in the ranking database\n",
    "    mtx = mtx[mtx.index.isin(genes_in_db)]\n",
    "    \n",
    "    return mtx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx = process_gse57872('GSE57872.mtx.txt', genes_in_db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MGH264_A01</th>\n",
       "      <th>MGH264_A02</th>\n",
       "      <th>MGH264_A03</th>\n",
       "      <th>MGH264_A04</th>\n",
       "      <th>MGH264_A05</th>\n",
       "      <th>MGH264_A06</th>\n",
       "      <th>MGH264_A07</th>\n",
       "      <th>MGH264_A08</th>\n",
       "      <th>MGH264_A10</th>\n",
       "      <th>MGH264_A11</th>\n",
       "      <th>...</th>\n",
       "      <th>MGH31_H02</th>\n",
       "      <th>MGH31_H04</th>\n",
       "      <th>MGH31_H05</th>\n",
       "      <th>MGH31_H06</th>\n",
       "      <th>MGH31_H07</th>\n",
       "      <th>MGH31_H08</th>\n",
       "      <th>MGH31_H09</th>\n",
       "      <th>MGH31_H10</th>\n",
       "      <th>MGH31_H11</th>\n",
       "      <th>MGH31_H12</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENE</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OSR1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KRT73</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPRR2A</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STATH</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HOXC9</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.070857</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 430 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        MGH264_A01  MGH264_A02  MGH264_A03  MGH264_A04  MGH264_A05  \\\n",
       "GENE                                                                 \n",
       "OSR1           0.0         0.0         0.0         0.0         0.0   \n",
       "KRT73          0.0         0.0         0.0         0.0         0.0   \n",
       "SPRR2A         0.0         0.0         0.0         0.0         0.0   \n",
       "STATH          0.0         0.0         0.0         0.0         0.0   \n",
       "HOXC9          0.0         0.0         0.0         0.0         0.0   \n",
       "\n",
       "        MGH264_A06  MGH264_A07  MGH264_A08  MGH264_A10  MGH264_A11    ...      \\\n",
       "GENE                                                                  ...       \n",
       "OSR1           0.0         0.0    0.000000         0.0         0.0    ...       \n",
       "KRT73          0.0         0.0    0.000000         0.0         0.0    ...       \n",
       "SPRR2A         0.0         0.0    0.000000         0.0         0.0    ...       \n",
       "STATH          0.0         0.0    0.000000         0.0         0.0    ...       \n",
       "HOXC9          0.0         0.0    2.070857         0.0         0.0    ...       \n",
       "\n",
       "        MGH31_H02  MGH31_H04  MGH31_H05  MGH31_H06  MGH31_H07  MGH31_H08  \\\n",
       "GENE                                                                       \n",
       "OSR1          0.0        0.0        0.0        0.0        0.0        0.0   \n",
       "KRT73         0.0        0.0        0.0        0.0        0.0        0.0   \n",
       "SPRR2A        0.0        0.0        0.0        0.0        0.0        0.0   \n",
       "STATH         0.0        0.0        0.0        0.0        0.0        0.0   \n",
       "HOXC9         0.0        0.0        0.0        0.0        0.0        0.0   \n",
       "\n",
       "        MGH31_H09  MGH31_H10  MGH31_H11  MGH31_H12  \n",
       "GENE                                                \n",
       "OSR1          0.0        0.0        0.0        0.0  \n",
       "KRT73         0.0        0.0        0.0        0.0  \n",
       "SPRR2A        0.0        0.0        0.0        0.0  \n",
       "STATH         0.0        0.0        0.0        0.0  \n",
       "HOXC9         0.0        0.0        0.0        0.0  \n",
       "\n",
       "[5 rows x 430 columns]"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mtx.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5753, 430)"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mtx.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "452"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(mtx.index).intersection(set(human_tfs)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx.to_csv(os.path.join(RESULTS_FOLDER, 'GSE57872.mtx.tsv'), sep='\\t', index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GSE76312 - CML"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_gse76312(fname, genes_in_db):\n",
    "    # Load CSV file\n",
    "    mtx = pd.read_csv(os.path.join(RESOURCES_FOLDER, fname), sep='\\t', index_col=0)\n",
    "    \n",
    "    # Remove duplicate gene symbols (keep last when sorted ascending order according to row sum)\n",
    "    mtx = mtx.iloc[mtx.sum(axis=1).argsort()]\n",
    "    mtx = mtx[~mtx.index.duplicated(keep='last')]\n",
    "    \n",
    "    ## From FPKM to log2(TPM/10+1)\n",
    "    mtx = (mtx / mtx.sum(axis=0)) * 10.0e6\n",
    "    mtx = mtx.applymap(lambda tpm: math.log2((tpm/10.0) + 1.0))\n",
    "    \n",
    "    # Keep only genes that are in the ranking database\n",
    "    mtx = mtx[mtx.index.isin(genes_in_db)]\n",
    "    \n",
    "    return mtx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx = process_gse76312('GSE76312.mtx.txt', genes_in_db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OX2046D6</th>\n",
       "      <th>OX2046E12</th>\n",
       "      <th>OX2046D11</th>\n",
       "      <th>OX2046B5</th>\n",
       "      <th>OX2046H5</th>\n",
       "      <th>OX2046E9</th>\n",
       "      <th>OX2046D8</th>\n",
       "      <th>OX2046E7</th>\n",
       "      <th>OX2046H12</th>\n",
       "      <th>OX2046H9</th>\n",
       "      <th>...</th>\n",
       "      <th>CML15_3mTKIC7</th>\n",
       "      <th>CML15_3mTKIB10</th>\n",
       "      <th>CML15_3mTKIA6</th>\n",
       "      <th>CML15_3mTKIB11</th>\n",
       "      <th>CML15_3mTKID6</th>\n",
       "      <th>CML15_3mTKIC9</th>\n",
       "      <th>CML15_3mTKIC1</th>\n",
       "      <th>CML15_3mTKIC8</th>\n",
       "      <th>CML15_3mTKID9</th>\n",
       "      <th>CML15_3mTKID10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PPY</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCDC151</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR105-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ALX1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR101-2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 2287 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          OX2046D6  OX2046E12  OX2046D11  OX2046B5  OX2046H5  OX2046E9  \\\n",
       "PPY            0.0        0.0        0.0       0.0       0.0       0.0   \n",
       "CCDC151        0.0        0.0        0.0       0.0       0.0       0.0   \n",
       "MIR105-1       0.0        0.0        0.0       0.0       0.0       0.0   \n",
       "ALX1           0.0        0.0        0.0       0.0       0.0       0.0   \n",
       "MIR101-2       0.0        0.0        0.0       0.0       0.0       0.0   \n",
       "\n",
       "          OX2046D8  OX2046E7  OX2046H12  OX2046H9       ...        \\\n",
       "PPY            0.0       0.0        0.0       0.0       ...         \n",
       "CCDC151        0.0       0.0        0.0       0.0       ...         \n",
       "MIR105-1       0.0       0.0        0.0       0.0       ...         \n",
       "ALX1           0.0       0.0        0.0       0.0       ...         \n",
       "MIR101-2       0.0       0.0        0.0       0.0       ...         \n",
       "\n",
       "          CML15_3mTKIC7  CML15_3mTKIB10  CML15_3mTKIA6  CML15_3mTKIB11  \\\n",
       "PPY                 0.0             0.0            0.0             0.0   \n",
       "CCDC151             0.0             0.0            0.0             0.0   \n",
       "MIR105-1            0.0             0.0            0.0             0.0   \n",
       "ALX1                0.0             0.0            0.0             0.0   \n",
       "MIR101-2            0.0             0.0            0.0             0.0   \n",
       "\n",
       "          CML15_3mTKID6  CML15_3mTKIC9  CML15_3mTKIC1  CML15_3mTKIC8  \\\n",
       "PPY                 0.0            0.0            0.0            0.0   \n",
       "CCDC151             0.0            0.0            0.0            0.0   \n",
       "MIR105-1            0.0            0.0            0.0            0.0   \n",
       "ALX1                0.0            0.0            0.0            0.0   \n",
       "MIR101-2            0.0            0.0            0.0            0.0   \n",
       "\n",
       "          CML15_3mTKID9  CML15_3mTKID10  \n",
       "PPY                 0.0             0.0  \n",
       "CCDC151             0.0             0.0  \n",
       "MIR105-1            0.0             0.0  \n",
       "ALX1                0.0             0.0  \n",
       "MIR101-2            0.0             0.0  \n",
       "\n",
       "[5 rows x 2287 columns]"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mtx.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(18701, 2287)"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mtx.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1527"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(mtx.index).intersection(set(human_tfs)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx.to_csv(os.path.join(RESULTS_FOLDER, 'GSE76312.mtx.tsv'), sep='\\t', index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GSE75688 - Breast cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_gse75688(fname, genes_in_db):\n",
    "    # Load CSV file\n",
    "    mtx = pd.read_csv(os.path.join(RESOURCES_FOLDER, fname), sep='\\t', index_col=1)\n",
    "    del mtx['gene_id']\n",
    "    del mtx ['gene_type']\n",
    "\n",
    "    # Remove duplicate gene symbols (keep last when sorted ascending order according to row sum)\n",
    "    mtx = mtx.iloc[mtx.sum(axis=1).argsort()]\n",
    "    mtx = mtx[~mtx.index.duplicated(keep='last')]\n",
    "    \n",
    "    # From FPKM to log2(TPM/10+1)\n",
    "    mtx = mtx.applymap(lambda tpm: math.log2((tpm/10.0) + 1.0))\n",
    "    \n",
    "    # Keep only genes that are in the ranking database\n",
    "    mtx = mtx[mtx.index.isin(genes_in_db)]\n",
    "    \n",
    "    return mtx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx = process_gse75688('GSE75688.mtx.txt', genes_in_db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BC01_Pooled</th>\n",
       "      <th>BC01_Tumor</th>\n",
       "      <th>BC02_Pooled</th>\n",
       "      <th>BC03_Pooled</th>\n",
       "      <th>BC03LN_Pooled</th>\n",
       "      <th>BC04_Pooled</th>\n",
       "      <th>BC05_Pooled</th>\n",
       "      <th>BC06_Pooled</th>\n",
       "      <th>BC07_Tumor</th>\n",
       "      <th>BC07LN_Pooled</th>\n",
       "      <th>...</th>\n",
       "      <th>BC11_04</th>\n",
       "      <th>BC11_07</th>\n",
       "      <th>BC11_28</th>\n",
       "      <th>BC11_43</th>\n",
       "      <th>BC11_56</th>\n",
       "      <th>BC11_69</th>\n",
       "      <th>BC11_70</th>\n",
       "      <th>BC11_78</th>\n",
       "      <th>BC11_81</th>\n",
       "      <th>BC11_88</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SNORD119</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDY1B</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR5A2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR5B12</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR5B2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 563 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           BC01_Pooled  BC01_Tumor  BC02_Pooled  BC03_Pooled  BC03LN_Pooled  \\\n",
       "gene_name                                                                     \n",
       "SNORD119           0.0         0.0          0.0          0.0            0.0   \n",
       "CDY1B              0.0         0.0          0.0          0.0            0.0   \n",
       "OR5A2              0.0         0.0          0.0          0.0            0.0   \n",
       "OR5B12             0.0         0.0          0.0          0.0            0.0   \n",
       "OR5B2              0.0         0.0          0.0          0.0            0.0   \n",
       "\n",
       "           BC04_Pooled  BC05_Pooled  BC06_Pooled  BC07_Tumor  BC07LN_Pooled  \\\n",
       "gene_name                                                                     \n",
       "SNORD119           0.0          0.0          0.0         0.0            0.0   \n",
       "CDY1B              0.0          0.0          0.0         0.0            0.0   \n",
       "OR5A2              0.0          0.0          0.0         0.0            0.0   \n",
       "OR5B12             0.0          0.0          0.0         0.0            0.0   \n",
       "OR5B2              0.0          0.0          0.0         0.0            0.0   \n",
       "\n",
       "            ...     BC11_04  BC11_07  BC11_28  BC11_43  BC11_56  BC11_69  \\\n",
       "gene_name   ...                                                            \n",
       "SNORD119    ...         0.0      0.0      0.0      0.0      0.0      0.0   \n",
       "CDY1B       ...         0.0      0.0      0.0      0.0      0.0      0.0   \n",
       "OR5A2       ...         0.0      0.0      0.0      0.0      0.0      0.0   \n",
       "OR5B12      ...         0.0      0.0      0.0      0.0      0.0      0.0   \n",
       "OR5B2       ...         0.0      0.0      0.0      0.0      0.0      0.0   \n",
       "\n",
       "           BC11_70  BC11_78  BC11_81  BC11_88  \n",
       "gene_name                                      \n",
       "SNORD119       0.0      0.0      0.0      0.0  \n",
       "CDY1B          0.0      0.0      0.0      0.0  \n",
       "OR5A2          0.0      0.0      0.0      0.0  \n",
       "OR5B12         0.0      0.0      0.0      0.0  \n",
       "OR5B2          0.0      0.0      0.0      0.0  \n",
       "\n",
       "[5 rows x 563 columns]"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mtx.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(19685, 563)"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mtx.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1565"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(mtx.index).intersection(set(human_tfs)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx.to_csv(os.path.join(RESULTS_FOLDER, 'GSE75688.mtx.tsv'), sep='\\t', index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GSE69405 - Lung carcinoma"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_gse69405(fname, genes_in_db):\n",
    "    # Load CSV file\n",
    "    mtx = pd.read_csv(os.path.join(RESOURCES_FOLDER, fname), sep='\\t', index_col=1)\n",
    "    del mtx['gene_id']\n",
    "    del mtx ['gene_type']\n",
    "    \n",
    "    # Remove duplicate gene symbols (keep last when sorted ascending order according to row sum)\n",
    "    mtx = mtx.iloc[mtx.sum(axis=1).argsort()]\n",
    "    mtx = mtx[~mtx.index.duplicated(keep='last')]\n",
    "    \n",
    "    # From FPKM to log2(TPM/10+1)\n",
    "    mtx = mtx.applymap(lambda tpm: math.log2((tpm/10.0) + 1.0))\n",
    "    \n",
    "    # Keep only genes that are in the ranking database\n",
    "    mtx = mtx[mtx.index.isin(genes_in_db)]\n",
    "    \n",
    "    return mtx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx = process_gse69405('GSE69405.mtx.txt', genes_in_db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H358_Pooled</th>\n",
       "      <th>H358_SC05</th>\n",
       "      <th>H358_SC09</th>\n",
       "      <th>H358_SC10</th>\n",
       "      <th>H358_SC11</th>\n",
       "      <th>H358_SC12</th>\n",
       "      <th>H358_SC14</th>\n",
       "      <th>H358_SC16</th>\n",
       "      <th>H358_SC17</th>\n",
       "      <th>H358_SC18</th>\n",
       "      <th>...</th>\n",
       "      <th>LC-MBT-15_SC87</th>\n",
       "      <th>LC-MBT-15_SC89</th>\n",
       "      <th>LC-MBT-15_SC92</th>\n",
       "      <th>LC-MBT-15_SC94</th>\n",
       "      <th>LC-MBT-15_SC95</th>\n",
       "      <th>LC-MBT-15_SC96</th>\n",
       "      <th>LC-PT-45-mock</th>\n",
       "      <th>LC-PT-45-Selumetinib_R0</th>\n",
       "      <th>LC-PT-45-Selumetinib_R3</th>\n",
       "      <th>LC-PT-45-Selumetinib_R7</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR764</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR573</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNORA36B</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGK2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR10J1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 201 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           H358_Pooled  H358_SC05  H358_SC09  H358_SC10  H358_SC11  H358_SC12  \\\n",
       "gene_name                                                                       \n",
       "MIR764             0.0        0.0        0.0        0.0        0.0        0.0   \n",
       "MIR573             0.0        0.0        0.0        0.0        0.0        0.0   \n",
       "SNORA36B           0.0        0.0        0.0        0.0        0.0        0.0   \n",
       "PGK2               0.0        0.0        0.0        0.0        0.0        0.0   \n",
       "OR10J1             0.0        0.0        0.0        0.0        0.0        0.0   \n",
       "\n",
       "           H358_SC14  H358_SC16  H358_SC17  H358_SC18  \\\n",
       "gene_name                                               \n",
       "MIR764           0.0        0.0        0.0        0.0   \n",
       "MIR573           0.0        0.0        0.0        0.0   \n",
       "SNORA36B         0.0        0.0        0.0        0.0   \n",
       "PGK2             0.0        0.0        0.0        0.0   \n",
       "OR10J1           0.0        0.0        0.0        0.0   \n",
       "\n",
       "                    ...             LC-MBT-15_SC87  LC-MBT-15_SC89  \\\n",
       "gene_name           ...                                              \n",
       "MIR764              ...                        0.0             0.0   \n",
       "MIR573              ...                        0.0             0.0   \n",
       "SNORA36B            ...                        0.0             0.0   \n",
       "PGK2                ...                        0.0             0.0   \n",
       "OR10J1              ...                        0.0             0.0   \n",
       "\n",
       "           LC-MBT-15_SC92  LC-MBT-15_SC94  LC-MBT-15_SC95  LC-MBT-15_SC96  \\\n",
       "gene_name                                                                   \n",
       "MIR764                0.0             0.0             0.0             0.0   \n",
       "MIR573                0.0             0.0             0.0             0.0   \n",
       "SNORA36B              0.0             0.0             0.0             0.0   \n",
       "PGK2                  0.0             0.0             0.0             0.0   \n",
       "OR10J1                0.0             0.0             0.0             0.0   \n",
       "\n",
       "           LC-PT-45-mock  LC-PT-45-Selumetinib_R0  LC-PT-45-Selumetinib_R3  \\\n",
       "gene_name                                                                    \n",
       "MIR764               0.0                      0.0                      0.0   \n",
       "MIR573               0.0                      0.0                      0.0   \n",
       "SNORA36B             0.0                      0.0                      0.0   \n",
       "PGK2                 0.0                      0.0                      0.0   \n",
       "OR10J1               0.0                      0.0                      0.0   \n",
       "\n",
       "           LC-PT-45-Selumetinib_R7  \n",
       "gene_name                           \n",
       "MIR764                         0.0  \n",
       "MIR573                         0.0  \n",
       "SNORA36B                       0.0  \n",
       "PGK2                           0.0  \n",
       "OR10J1                         0.0  \n",
       "\n",
       "[5 rows x 201 columns]"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mtx.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(19659, 201)"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mtx.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1565"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(mtx.index).intersection(set(human_tfs)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx.to_csv(os.path.join(RESULTS_FOLDER, 'GSE69405.mtx.tsv'), sep='\\t', index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GSE81861 - Colorectal tumor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_gse81861(fname, genes_in_db):\n",
    "    # Load CSV file\n",
    "    mtx = pd.read_csv(os.path.join(RESOURCES_FOLDER, fname), sep=',', index_col=0)\n",
    "    \n",
    "    # Extract gene symbol\n",
    "    def gene(s):\n",
    "        parts = s.split('_')\n",
    "        return \"_\".join(parts[1:-1])\n",
    "    mtx.index = list(map(gene, mtx.index))\n",
    "    \n",
    "    # Remove duplicate gene symbols (keep last when sorted ascending order according to row sum)\n",
    "    mtx = mtx.iloc[mtx.sum(axis=1).argsort()]\n",
    "    mtx = mtx[~mtx.index.duplicated(keep='last')]\n",
    "    \n",
    "    # From FPKM to log2(TPM/10+1)\n",
    "    mtx = (mtx / mtx.sum(axis=0)) * 10.0e6\n",
    "    mtx = mtx.applymap(lambda tpm: math.log2((tpm/10.0) + 1.0))\n",
    "    \n",
    "    # Keep only genes that are in the ranking database\n",
    "    mtx = mtx[mtx.index.isin(genes_in_db)]\n",
    "    \n",
    "    return mtx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "process_gse81861 = partial(process_gse81861, genes_in_db=genes_in_db)\n",
    "mtx = pd.concat([\n",
    "    process_gse81861('GSE81861_CRC_NM_all_cells_FPKM.csv'),\n",
    "    process_gse81861('GSE81861_CRC_NM_epithelial_cells_FPKM.csv'),\n",
    "    process_gse81861('GSE81861_CRC_tumor_all_cells_FPKM.csv'),\n",
    "    process_gse81861('GSE81861_CRC_tumor_epithelial_cells_FPKM.csv'),\n",
    "    process_gse81861('GSE81861_Cell_Line_FPKM.csv'),\n",
    "    ], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(19683, 1583)"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mtx.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RHC3934__Bcell__#7DEA7B</th>\n",
       "      <th>RHC3944__Bcell__#7DEA7B</th>\n",
       "      <th>RHC3962__Tcell__#C6E879</th>\n",
       "      <th>RHC4003__Bcell__#7DEA7B</th>\n",
       "      <th>RHC4005__Bcell__#7DEA7B</th>\n",
       "      <th>RHC4007__Bcell__#7DEA7B</th>\n",
       "      <th>RHC4008__NA__grey</th>\n",
       "      <th>RHC4220__Bcell__#7DEA7B</th>\n",
       "      <th>RHC4227__Epithelial__#2749FE</th>\n",
       "      <th>RHC4240__Epithelial__#2749FE</th>\n",
       "      <th>...</th>\n",
       "      <th>RHC2497__H1_B2__brown</th>\n",
       "      <th>RHC2498__H1_B2__brown</th>\n",
       "      <th>RHC2499__H1_B2__brown</th>\n",
       "      <th>RHC2500__H1_B2__brown</th>\n",
       "      <th>RHC2501__H1_B2__brown</th>\n",
       "      <th>RHC2502__H1_B2__brown</th>\n",
       "      <th>RHC2503__H1_B2__brown</th>\n",
       "      <th>RHC2504__H1_B2__brown</th>\n",
       "      <th>RHC2505__H1_B2__brown</th>\n",
       "      <th>RHC2506__H1_B2__brown</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>3.358942</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.735678</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.314749</td>\n",
       "      <td>6.163942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.105035</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.206954</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.558213</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.195409</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.664019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.083355</td>\n",
       "      <td>0.115427</td>\n",
       "      <td>0.413562</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.040125</td>\n",
       "      <td>0.053906</td>\n",
       "      <td>0.665874</td>\n",
       "      <td>2.608487</td>\n",
       "      <td>1.393939</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GALT</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.488154</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.260969</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.893560</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1583 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        RHC3934__Bcell__#7DEA7B  RHC3944__Bcell__#7DEA7B  \\\n",
       "A1BG                        0.0                      0.0   \n",
       "A1CF                        0.0                      0.0   \n",
       "A2M                         0.0                      0.0   \n",
       "A2ML1                       0.0                      0.0   \n",
       "A4GALT                      0.0                      0.0   \n",
       "\n",
       "        RHC3962__Tcell__#C6E879  RHC4003__Bcell__#7DEA7B  \\\n",
       "A1BG                   0.000000                 0.000000   \n",
       "A1CF                   0.105035                 0.000000   \n",
       "A2M                    0.000000                 0.000000   \n",
       "A2ML1                  0.000000                 0.000000   \n",
       "A4GALT                 0.000000                 7.488154   \n",
       "\n",
       "        RHC4005__Bcell__#7DEA7B  RHC4007__Bcell__#7DEA7B  RHC4008__NA__grey  \\\n",
       "A1BG                        0.0                 0.000000                0.0   \n",
       "A1CF                        0.0                 0.206954                0.0   \n",
       "A2M                         0.0                 0.000000                0.0   \n",
       "A2ML1                       0.0                 0.000000                0.0   \n",
       "A4GALT                      0.0                 0.000000                0.0   \n",
       "\n",
       "        RHC4220__Bcell__#7DEA7B  RHC4227__Epithelial__#2749FE  \\\n",
       "A1BG                        0.0                           0.0   \n",
       "A1CF                        0.0                           0.0   \n",
       "A2M                         0.0                           0.0   \n",
       "A2ML1                       0.0                           0.0   \n",
       "A4GALT                      0.0                           0.0   \n",
       "\n",
       "        RHC4240__Epithelial__#2749FE          ...            \\\n",
       "A1BG                        0.000000          ...             \n",
       "A1CF                        6.558213          ...             \n",
       "A2M                         0.000000          ...             \n",
       "A2ML1                       0.000000          ...             \n",
       "A4GALT                      0.000000          ...             \n",
       "\n",
       "        RHC2497__H1_B2__brown  RHC2498__H1_B2__brown  RHC2499__H1_B2__brown  \\\n",
       "A1BG                 3.358942               0.000000               0.000000   \n",
       "A1CF                 0.000000               0.000000               0.000000   \n",
       "A2M                  0.000000               0.000000               0.000000   \n",
       "A2ML1                0.083355               0.115427               0.413562   \n",
       "A4GALT               0.000000               0.000000               0.000000   \n",
       "\n",
       "        RHC2500__H1_B2__brown  RHC2501__H1_B2__brown  RHC2502__H1_B2__brown  \\\n",
       "A1BG                      0.0               0.000000               0.000000   \n",
       "A1CF                      0.0               0.000000               0.000000   \n",
       "A2M                       0.0               0.000000               0.000000   \n",
       "A2ML1                     0.0               4.040125               0.053906   \n",
       "A4GALT                    0.0               1.260969               0.000000   \n",
       "\n",
       "        RHC2503__H1_B2__brown  RHC2504__H1_B2__brown  RHC2505__H1_B2__brown  \\\n",
       "A1BG                 1.735678               0.000000               1.314749   \n",
       "A1CF                 0.000000               0.000000               0.000000   \n",
       "A2M                  3.195409               0.000000               0.000000   \n",
       "A2ML1                0.665874               2.608487               1.393939   \n",
       "A4GALT               0.000000               0.000000               0.893560   \n",
       "\n",
       "        RHC2506__H1_B2__brown  \n",
       "A1BG                 6.163942  \n",
       "A1CF                 0.000000  \n",
       "A2M                  1.664019  \n",
       "A2ML1                0.000000  \n",
       "A4GALT               0.000000  \n",
       "\n",
       "[5 rows x 1583 columns]"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mtx.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1565"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(mtx.index).intersection(set(human_tfs)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx.to_csv(os.path.join(RESULTS_FOLDER, 'GSE81861.mtx.tsv'), sep='\\t', index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GSE73121 - RCC\n",
    "\n",
    "Some difficulties with this dataset:\n",
    "1. TPM-values need to be transformed, log2(TPM/10+1). EASY\n",
    "2. Separate file for each cell in the experiment. EASY\n",
    "3. The genes are specified as ensemble gene IDs (ENSG) and need to be converted to HGNC symbols. DIFFICULT\n",
    "\n",
    "Downloading FPKM table from SCPortalen is not a solution becuase genes are also provided as ENSG numbers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GSM1887215_PDX_mRCC_Pooled.TPM.txt GSM1887276_PDX_pRCC_SC_57.TPM.txt\n",
      "GSM1887216_PDX_mRCC_SC_01.TPM.txt  GSM1887277_PDX_pRCC_SC_58.TPM.txt\n",
      "GSM1887217_PDX_mRCC_SC_03.TPM.txt  GSM1887278_PDX_pRCC_SC_59.TPM.txt\n",
      "GSM1887218_PDX_mRCC_SC_04.TPM.txt  GSM1887279_PDX_pRCC_SC_60.TPM.txt\n",
      "GSM1887219_PDX_mRCC_SC_05.TPM.txt  GSM1887280_PDX_pRCC_SC_61.TPM.txt\n",
      "GSM1887220_PDX_mRCC_SC_06.TPM.txt  GSM1887281_PDX_pRCC_SC_62.TPM.txt\n",
      "GSM1887221_PDX_mRCC_SC_08.TPM.txt  GSM1887282_PDX_pRCC_SC_65.TPM.txt\n",
      "GSM1887222_PDX_mRCC_SC_12.TPM.txt  GSM1887283_PDX_pRCC_SC_66.TPM.txt\n",
      "GSM1887223_PDX_mRCC_SC_14.TPM.txt  GSM1887284_PDX_pRCC_SC_67.TPM.txt\n",
      "GSM1887224_PDX_mRCC_SC_15.TPM.txt  GSM1887285_PDX_pRCC_SC_68.TPM.txt\n",
      "GSM1887225_PDX_mRCC_SC_20.TPM.txt  GSM1887286_PDX_pRCC_SC_70.TPM.txt\n",
      "GSM1887226_PDX_mRCC_SC_21.TPM.txt  GSM1887287_PDX_pRCC_SC_72.TPM.txt\n",
      "GSM1887227_PDX_mRCC_SC_34.TPM.txt  GSM1887288_PDX_pRCC_SC_73.TPM.txt\n",
      "GSM1887228_PDX_mRCC_SC_41.TPM.txt  GSM1887289_PDX_pRCC_SC_75.TPM.txt\n",
      "GSM1887229_PDX_mRCC_SC_44.TPM.txt  GSM1887290_PDX_pRCC_SC_78.TPM.txt\n",
      "GSM1887230_PDX_mRCC_SC_52.TPM.txt  GSM1887291_PDX_pRCC_SC_79.TPM.txt\n",
      "GSM1887231_PDX_mRCC_SC_56.TPM.txt  GSM1887292_PDX_pRCC_SC_84.TPM.txt\n",
      "GSM1887232_PDX_mRCC_SC_60.TPM.txt  GSM1887293_PDX_pRCC_SC_87.TPM.txt\n",
      "GSM1887233_PDX_mRCC_SC_61.TPM.txt  GSM1887294_PDX_pRCC_SC_88.TPM.txt\n",
      "GSM1887234_PDX_mRCC_SC_63.TPM.txt  GSM1887295_PDX_pRCC_SC_89.TPM.txt\n",
      "GSM1887235_PDX_mRCC_SC_64.TPM.txt  GSM1887296_PDX_pRCC_SC_90.TPM.txt\n",
      "GSM1887236_PDX_mRCC_SC_68.TPM.txt  GSM1887297_PDX_pRCC_SC_91.TPM.txt\n",
      "GSM1887237_PDX_mRCC_SC_69.TPM.txt  GSM1887299_PDX_pRCC_SC_93.TPM.txt\n",
      "GSM1887238_PDX_mRCC_SC_70.TPM.txt  GSM1887301_PDX_pRCC_SC_94.TPM.txt\n",
      "GSM1887239_PDX_mRCC_SC_77.TPM.txt  GSM1887303_Pt_mRCC_Pooled.TPM.txt\n",
      "GSM1887240_PDX_mRCC_SC_79.TPM.txt  GSM1887306_Pt_mRCC_SC_01.TPM.txt\n",
      "GSM1887241_PDX_mRCC_SC_80.TPM.txt  GSM1887308_Pt_mRCC_SC_03.TPM.txt\n",
      "GSM1887242_PDX_mRCC_SC_81.TPM.txt  GSM1887310_Pt_mRCC_SC_05.TPM.txt\n",
      "GSM1887243_PDX_mRCC_SC_86.TPM.txt  GSM1887312_Pt_mRCC_SC_10.TPM.txt\n",
      "GSM1887244_PDX_mRCC_SC_87.TPM.txt  GSM1887315_Pt_mRCC_SC_11.TPM.txt\n",
      "GSM1887245_PDX_mRCC_SC_88.TPM.txt  GSM1887316_Pt_mRCC_SC_12.TPM.txt\n",
      "GSM1887246_PDX_mRCC_SC_89.TPM.txt  GSM1887318_Pt_mRCC_SC_17.TPM.txt\n",
      "GSM1887247_PDX_mRCC_SC_90.TPM.txt  GSM1887321_Pt_mRCC_SC_21.TPM.txt\n",
      "GSM1887248_PDX_mRCC_SC_92.TPM.txt  GSM1887323_Pt_mRCC_SC_23.TPM.txt\n",
      "GSM1887249_PDX_mRCC_SC_93.TPM.txt  GSM1887325_Pt_mRCC_SC_24.TPM.txt\n",
      "GSM1887250_PDX_mRCC_SC_94.TPM.txt  GSM1887327_Pt_mRCC_SC_30.TPM.txt\n",
      "GSM1887251_PDX_mRCC_SC_95.TPM.txt  GSM1887329_Pt_mRCC_SC_33.TPM.txt\n",
      "GSM1887252_PDX_pRCC_Pooled.TPM.txt GSM1887330_Pt_mRCC_SC_36.TPM.txt\n",
      "GSM1887253_PDX_pRCC_SC_03.TPM.txt  GSM1887331_Pt_mRCC_SC_39.TPM.txt\n",
      "GSM1887254_PDX_pRCC_SC_05.TPM.txt  GSM1887332_Pt_mRCC_SC_43.TPM.txt\n",
      "GSM1887255_PDX_pRCC_SC_06.TPM.txt  GSM1887333_Pt_mRCC_SC_53.TPM.txt\n",
      "GSM1887256_PDX_pRCC_SC_10.TPM.txt  GSM1887334_Pt_mRCC_SC_56.TPM.txt\n",
      "GSM1887257_PDX_pRCC_SC_13.TPM.txt  GSM1887335_Pt_mRCC_SC_57.TPM.txt\n",
      "GSM1887258_PDX_pRCC_SC_14.TPM.txt  GSM1887336_Pt_mRCC_SC_58.TPM.txt\n",
      "GSM1887259_PDX_pRCC_SC_16.TPM.txt  GSM1887337_Pt_mRCC_SC_59.TPM.txt\n",
      "GSM1887260_PDX_pRCC_SC_24.TPM.txt  GSM1887338_Pt_mRCC_SC_61.TPM.txt\n",
      "GSM1887261_PDX_pRCC_SC_27.TPM.txt  GSM1887339_Pt_mRCC_SC_64.TPM.txt\n",
      "GSM1887262_PDX_pRCC_SC_28.TPM.txt  GSM1887340_Pt_mRCC_SC_69.TPM.txt\n",
      "GSM1887263_PDX_pRCC_SC_29.TPM.txt  GSM1887341_Pt_mRCC_SC_71.TPM.txt\n",
      "GSM1887264_PDX_pRCC_SC_34.TPM.txt  GSM1887342_Pt_mRCC_SC_75.TPM.txt\n",
      "GSM1887265_PDX_pRCC_SC_35.TPM.txt  GSM1887343_Pt_mRCC_SC_79.TPM.txt\n",
      "GSM1887266_PDX_pRCC_SC_36.TPM.txt  GSM1887344_Pt_mRCC_SC_81.TPM.txt\n",
      "GSM1887267_PDX_pRCC_SC_37.TPM.txt  GSM1887345_Pt_mRCC_SC_84.TPM.txt\n",
      "GSM1887268_PDX_pRCC_SC_46.TPM.txt  GSM1887346_Pt_mRCC_SC_85.TPM.txt\n",
      "GSM1887269_PDX_pRCC_SC_49.TPM.txt  GSM1887347_Pt_mRCC_SC_86.TPM.txt\n",
      "GSM1887270_PDX_pRCC_SC_50.TPM.txt  GSM1887348_Pt_mRCC_SC_87.TPM.txt\n",
      "GSM1887271_PDX_pRCC_SC_51.TPM.txt  GSM1887349_Pt_mRCC_SC_88.TPM.txt\n",
      "GSM1887272_PDX_pRCC_SC_52.TPM.txt  GSM1887350_Pt_mRCC_SC_90.TPM.txt\n",
      "GSM1887273_PDX_pRCC_SC_53.TPM.txt  GSM1887351_Pt_mRCC_SC_92.TPM.txt\n",
      "GSM1887274_PDX_pRCC_SC_55.TPM.txt  GSM1887352_Pt_mRCC_SC_93.TPM.txt\n",
      "GSM1887275_PDX_pRCC_SC_56.TPM.txt\n"
     ]
    }
   ],
   "source": [
    "!ls {RESOURCES_FOLDER}/GSE73121/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load(fname):\n",
    "    col = pd.read_csv(fname, sep='\\t', index_col=0, usecols=[0,5])\n",
    "    col.rename(columns={'TPM': os.path.basename(fname).split('.')[0]}, inplace=True)\n",
    "    return col.applymap(lambda tpm: math.log2((tpm/10.0) + 1.0))\n",
    "\n",
    "def process_gse73121(foldername, genes_in_db):\n",
    "    fnames = glob.glob(os.path.join(foldername, \"GSM*.TPM.txt\"))\n",
    "    return pd.concat([ load(fname) for fname in fnames], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx = process_gse73121(os.path.join(RESOURCES_FOLDER, \"GSE73121\"), genes_in_db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(57820, 121)"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mtx.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1887287_PDX_pRCC_SC_72</th>\n",
       "      <th>GSM1887294_PDX_pRCC_SC_88</th>\n",
       "      <th>GSM1887230_PDX_mRCC_SC_52</th>\n",
       "      <th>GSM1887285_PDX_pRCC_SC_68</th>\n",
       "      <th>GSM1887254_PDX_pRCC_SC_05</th>\n",
       "      <th>GSM1887274_PDX_pRCC_SC_55</th>\n",
       "      <th>GSM1887335_Pt_mRCC_SC_57</th>\n",
       "      <th>GSM1887336_Pt_mRCC_SC_58</th>\n",
       "      <th>GSM1887286_PDX_pRCC_SC_70</th>\n",
       "      <th>GSM1887345_Pt_mRCC_SC_84</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1887225_PDX_mRCC_SC_20</th>\n",
       "      <th>GSM1887218_PDX_mRCC_SC_04</th>\n",
       "      <th>GSM1887321_Pt_mRCC_SC_21</th>\n",
       "      <th>GSM1887261_PDX_pRCC_SC_27</th>\n",
       "      <th>GSM1887277_PDX_pRCC_SC_58</th>\n",
       "      <th>GSM1887223_PDX_mRCC_SC_14</th>\n",
       "      <th>GSM1887312_Pt_mRCC_SC_10</th>\n",
       "      <th>GSM1887340_Pt_mRCC_SC_69</th>\n",
       "      <th>GSM1887244_PDX_mRCC_SC_87</th>\n",
       "      <th>GSM1887264_PDX_pRCC_SC_34</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003.10</th>\n",
       "      <td>0.166073</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.954196</td>\n",
       "      <td>0.151859</td>\n",
       "      <td>1.529071</td>\n",
       "      <td>0.267835</td>\n",
       "      <td>1.216610</td>\n",
       "      <td>3.465322</td>\n",
       "      <td>1.226509</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.630405</td>\n",
       "      <td>3.762242</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.724214</td>\n",
       "      <td>2.931305</td>\n",
       "      <td>1.667665</td>\n",
       "      <td>0.465713</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.466758</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005.5</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419.8</th>\n",
       "      <td>3.853597</td>\n",
       "      <td>3.245648</td>\n",
       "      <td>4.774787</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.919531</td>\n",
       "      <td>4.191800</td>\n",
       "      <td>4.718910</td>\n",
       "      <td>3.408712</td>\n",
       "      <td>2.931305</td>\n",
       "      <td>4.652601</td>\n",
       "      <td>...</td>\n",
       "      <td>4.501375</td>\n",
       "      <td>5.600448</td>\n",
       "      <td>5.326573</td>\n",
       "      <td>4.217231</td>\n",
       "      <td>1.473008</td>\n",
       "      <td>4.737849</td>\n",
       "      <td>3.994761</td>\n",
       "      <td>0.084064</td>\n",
       "      <td>2.713256</td>\n",
       "      <td>2.81578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457.9</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.621056</td>\n",
       "      <td>2.010422</td>\n",
       "      <td>0.344828</td>\n",
       "      <td>1.937344</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.754460</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.189034</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.071763</td>\n",
       "      <td>0.516015</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014355</td>\n",
       "      <td>0.097611</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.048236</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460.12</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.075875</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 121 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    GSM1887287_PDX_pRCC_SC_72  GSM1887294_PDX_pRCC_SC_88  \\\n",
       "gene_id                                                                    \n",
       "ENSG00000000003.10                   0.166073                   0.000000   \n",
       "ENSG00000000005.5                    0.000000                   0.000000   \n",
       "ENSG00000000419.8                    3.853597                   3.245648   \n",
       "ENSG00000000457.9                    0.000000                   0.000000   \n",
       "ENSG00000000460.12                   0.000000                   0.000000   \n",
       "\n",
       "                    GSM1887230_PDX_mRCC_SC_52  GSM1887285_PDX_pRCC_SC_68  \\\n",
       "gene_id                                                                    \n",
       "ENSG00000000003.10                   1.954196                   0.151859   \n",
       "ENSG00000000005.5                    0.000000                   0.000000   \n",
       "ENSG00000000419.8                    4.774787                   0.000000   \n",
       "ENSG00000000457.9                    0.000000                   0.621056   \n",
       "ENSG00000000460.12                   0.000000                   0.000000   \n",
       "\n",
       "                    GSM1887254_PDX_pRCC_SC_05  GSM1887274_PDX_pRCC_SC_55  \\\n",
       "gene_id                                                                    \n",
       "ENSG00000000003.10                   1.529071                   0.267835   \n",
       "ENSG00000000005.5                    0.000000                   0.000000   \n",
       "ENSG00000000419.8                    2.919531                   4.191800   \n",
       "ENSG00000000457.9                    2.010422                   0.344828   \n",
       "ENSG00000000460.12                   0.000000                   0.000000   \n",
       "\n",
       "                    GSM1887335_Pt_mRCC_SC_57  GSM1887336_Pt_mRCC_SC_58  \\\n",
       "gene_id                                                                  \n",
       "ENSG00000000003.10                  1.216610                  3.465322   \n",
       "ENSG00000000005.5                   0.000000                  0.000000   \n",
       "ENSG00000000419.8                   4.718910                  3.408712   \n",
       "ENSG00000000457.9                   1.937344                  0.000000   \n",
       "ENSG00000000460.12                  1.075875                  0.000000   \n",
       "\n",
       "                    GSM1887286_PDX_pRCC_SC_70  GSM1887345_Pt_mRCC_SC_84  \\\n",
       "gene_id                                                                   \n",
       "ENSG00000000003.10                   1.226509                  0.000000   \n",
       "ENSG00000000005.5                    0.000000                  0.000000   \n",
       "ENSG00000000419.8                    2.931305                  4.652601   \n",
       "ENSG00000000457.9                    1.754460                  0.000000   \n",
       "ENSG00000000460.12                   0.000000                  0.000000   \n",
       "\n",
       "                              ...              GSM1887225_PDX_mRCC_SC_20  \\\n",
       "gene_id                       ...                                          \n",
       "ENSG00000000003.10            ...                               0.630405   \n",
       "ENSG00000000005.5             ...                               0.000000   \n",
       "ENSG00000000419.8             ...                               4.501375   \n",
       "ENSG00000000457.9             ...                               0.189034   \n",
       "ENSG00000000460.12            ...                               0.000000   \n",
       "\n",
       "                    GSM1887218_PDX_mRCC_SC_04  GSM1887321_Pt_mRCC_SC_21  \\\n",
       "gene_id                                                                   \n",
       "ENSG00000000003.10                   3.762242                  0.000000   \n",
       "ENSG00000000005.5                    0.000000                  0.000000   \n",
       "ENSG00000000419.8                    5.600448                  5.326573   \n",
       "ENSG00000000457.9                    0.000000                  0.071763   \n",
       "ENSG00000000460.12                   0.000000                  0.000000   \n",
       "\n",
       "                    GSM1887261_PDX_pRCC_SC_27  GSM1887277_PDX_pRCC_SC_58  \\\n",
       "gene_id                                                                    \n",
       "ENSG00000000003.10                   1.724214                   2.931305   \n",
       "ENSG00000000005.5                    0.000000                   0.000000   \n",
       "ENSG00000000419.8                    4.217231                   1.473008   \n",
       "ENSG00000000457.9                    0.516015                   0.000000   \n",
       "ENSG00000000460.12                   0.000000                   0.000000   \n",
       "\n",
       "                    GSM1887223_PDX_mRCC_SC_14  GSM1887312_Pt_mRCC_SC_10  \\\n",
       "gene_id                                                                   \n",
       "ENSG00000000003.10                   1.667665                  0.465713   \n",
       "ENSG00000000005.5                    0.000000                  0.000000   \n",
       "ENSG00000000419.8                    4.737849                  3.994761   \n",
       "ENSG00000000457.9                    0.014355                  0.097611   \n",
       "ENSG00000000460.12                   0.000000                  0.000000   \n",
       "\n",
       "                    GSM1887340_Pt_mRCC_SC_69  GSM1887244_PDX_mRCC_SC_87  \\\n",
       "gene_id                                                                   \n",
       "ENSG00000000003.10                  0.000000                   0.466758   \n",
       "ENSG00000000005.5                   0.000000                   0.000000   \n",
       "ENSG00000000419.8                   0.084064                   2.713256   \n",
       "ENSG00000000457.9                   0.000000                   0.048236   \n",
       "ENSG00000000460.12                  0.000000                   0.000000   \n",
       "\n",
       "                    GSM1887264_PDX_pRCC_SC_34  \n",
       "gene_id                                        \n",
       "ENSG00000000003.10                    0.00000  \n",
       "ENSG00000000005.5                     0.00000  \n",
       "ENSG00000000419.8                     2.81578  \n",
       "ENSG00000000457.9                     0.00000  \n",
       "ENSG00000000460.12                    0.00000  \n",
       "\n",
       "[5 rows x 121 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mtx.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx.index = list(map(lambda g: g.split('.')[0], mtx.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1887287_PDX_pRCC_SC_72</th>\n",
       "      <th>GSM1887294_PDX_pRCC_SC_88</th>\n",
       "      <th>GSM1887230_PDX_mRCC_SC_52</th>\n",
       "      <th>GSM1887285_PDX_pRCC_SC_68</th>\n",
       "      <th>GSM1887254_PDX_pRCC_SC_05</th>\n",
       "      <th>GSM1887274_PDX_pRCC_SC_55</th>\n",
       "      <th>GSM1887335_Pt_mRCC_SC_57</th>\n",
       "      <th>GSM1887336_Pt_mRCC_SC_58</th>\n",
       "      <th>GSM1887286_PDX_pRCC_SC_70</th>\n",
       "      <th>GSM1887345_Pt_mRCC_SC_84</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1887225_PDX_mRCC_SC_20</th>\n",
       "      <th>GSM1887218_PDX_mRCC_SC_04</th>\n",
       "      <th>GSM1887321_Pt_mRCC_SC_21</th>\n",
       "      <th>GSM1887261_PDX_pRCC_SC_27</th>\n",
       "      <th>GSM1887277_PDX_pRCC_SC_58</th>\n",
       "      <th>GSM1887223_PDX_mRCC_SC_14</th>\n",
       "      <th>GSM1887312_Pt_mRCC_SC_10</th>\n",
       "      <th>GSM1887340_Pt_mRCC_SC_69</th>\n",
       "      <th>GSM1887244_PDX_mRCC_SC_87</th>\n",
       "      <th>GSM1887264_PDX_pRCC_SC_34</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>0.166073</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.954196</td>\n",
       "      <td>0.151859</td>\n",
       "      <td>1.529071</td>\n",
       "      <td>0.267835</td>\n",
       "      <td>1.216610</td>\n",
       "      <td>3.465322</td>\n",
       "      <td>1.226509</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.630405</td>\n",
       "      <td>3.762242</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.724214</td>\n",
       "      <td>2.931305</td>\n",
       "      <td>1.667665</td>\n",
       "      <td>0.465713</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.466758</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>3.853597</td>\n",
       "      <td>3.245648</td>\n",
       "      <td>4.774787</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.919531</td>\n",
       "      <td>4.191800</td>\n",
       "      <td>4.718910</td>\n",
       "      <td>3.408712</td>\n",
       "      <td>2.931305</td>\n",
       "      <td>4.652601</td>\n",
       "      <td>...</td>\n",
       "      <td>4.501375</td>\n",
       "      <td>5.600448</td>\n",
       "      <td>5.326573</td>\n",
       "      <td>4.217231</td>\n",
       "      <td>1.473008</td>\n",
       "      <td>4.737849</td>\n",
       "      <td>3.994761</td>\n",
       "      <td>0.084064</td>\n",
       "      <td>2.713256</td>\n",
       "      <td>2.81578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.621056</td>\n",
       "      <td>2.010422</td>\n",
       "      <td>0.344828</td>\n",
       "      <td>1.937344</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.754460</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.189034</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.071763</td>\n",
       "      <td>0.516015</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014355</td>\n",
       "      <td>0.097611</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.048236</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.075875</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 121 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GSM1887287_PDX_pRCC_SC_72  GSM1887294_PDX_pRCC_SC_88  \\\n",
       "ENSG00000000003                   0.166073                   0.000000   \n",
       "ENSG00000000005                   0.000000                   0.000000   \n",
       "ENSG00000000419                   3.853597                   3.245648   \n",
       "ENSG00000000457                   0.000000                   0.000000   \n",
       "ENSG00000000460                   0.000000                   0.000000   \n",
       "\n",
       "                 GSM1887230_PDX_mRCC_SC_52  GSM1887285_PDX_pRCC_SC_68  \\\n",
       "ENSG00000000003                   1.954196                   0.151859   \n",
       "ENSG00000000005                   0.000000                   0.000000   \n",
       "ENSG00000000419                   4.774787                   0.000000   \n",
       "ENSG00000000457                   0.000000                   0.621056   \n",
       "ENSG00000000460                   0.000000                   0.000000   \n",
       "\n",
       "                 GSM1887254_PDX_pRCC_SC_05  GSM1887274_PDX_pRCC_SC_55  \\\n",
       "ENSG00000000003                   1.529071                   0.267835   \n",
       "ENSG00000000005                   0.000000                   0.000000   \n",
       "ENSG00000000419                   2.919531                   4.191800   \n",
       "ENSG00000000457                   2.010422                   0.344828   \n",
       "ENSG00000000460                   0.000000                   0.000000   \n",
       "\n",
       "                 GSM1887335_Pt_mRCC_SC_57  GSM1887336_Pt_mRCC_SC_58  \\\n",
       "ENSG00000000003                  1.216610                  3.465322   \n",
       "ENSG00000000005                  0.000000                  0.000000   \n",
       "ENSG00000000419                  4.718910                  3.408712   \n",
       "ENSG00000000457                  1.937344                  0.000000   \n",
       "ENSG00000000460                  1.075875                  0.000000   \n",
       "\n",
       "                 GSM1887286_PDX_pRCC_SC_70  GSM1887345_Pt_mRCC_SC_84  \\\n",
       "ENSG00000000003                   1.226509                  0.000000   \n",
       "ENSG00000000005                   0.000000                  0.000000   \n",
       "ENSG00000000419                   2.931305                  4.652601   \n",
       "ENSG00000000457                   1.754460                  0.000000   \n",
       "ENSG00000000460                   0.000000                  0.000000   \n",
       "\n",
       "                           ...              GSM1887225_PDX_mRCC_SC_20  \\\n",
       "ENSG00000000003            ...                               0.630405   \n",
       "ENSG00000000005            ...                               0.000000   \n",
       "ENSG00000000419            ...                               4.501375   \n",
       "ENSG00000000457            ...                               0.189034   \n",
       "ENSG00000000460            ...                               0.000000   \n",
       "\n",
       "                 GSM1887218_PDX_mRCC_SC_04  GSM1887321_Pt_mRCC_SC_21  \\\n",
       "ENSG00000000003                   3.762242                  0.000000   \n",
       "ENSG00000000005                   0.000000                  0.000000   \n",
       "ENSG00000000419                   5.600448                  5.326573   \n",
       "ENSG00000000457                   0.000000                  0.071763   \n",
       "ENSG00000000460                   0.000000                  0.000000   \n",
       "\n",
       "                 GSM1887261_PDX_pRCC_SC_27  GSM1887277_PDX_pRCC_SC_58  \\\n",
       "ENSG00000000003                   1.724214                   2.931305   \n",
       "ENSG00000000005                   0.000000                   0.000000   \n",
       "ENSG00000000419                   4.217231                   1.473008   \n",
       "ENSG00000000457                   0.516015                   0.000000   \n",
       "ENSG00000000460                   0.000000                   0.000000   \n",
       "\n",
       "                 GSM1887223_PDX_mRCC_SC_14  GSM1887312_Pt_mRCC_SC_10  \\\n",
       "ENSG00000000003                   1.667665                  0.465713   \n",
       "ENSG00000000005                   0.000000                  0.000000   \n",
       "ENSG00000000419                   4.737849                  3.994761   \n",
       "ENSG00000000457                   0.014355                  0.097611   \n",
       "ENSG00000000460                   0.000000                  0.000000   \n",
       "\n",
       "                 GSM1887340_Pt_mRCC_SC_69  GSM1887244_PDX_mRCC_SC_87  \\\n",
       "ENSG00000000003                  0.000000                   0.466758   \n",
       "ENSG00000000005                  0.000000                   0.000000   \n",
       "ENSG00000000419                  0.084064                   2.713256   \n",
       "ENSG00000000457                  0.000000                   0.048236   \n",
       "ENSG00000000460                  0.000000                   0.000000   \n",
       "\n",
       "                 GSM1887264_PDX_pRCC_SC_34  \n",
       "ENSG00000000003                    0.00000  \n",
       "ENSG00000000005                    0.00000  \n",
       "ENSG00000000419                    2.81578  \n",
       "ENSG00000000457                    0.00000  \n",
       "ENSG00000000460                    0.00000  \n",
       "\n",
       "[5 rows x 121 columns]"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mtx.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove duplicate gene symbols (keep last when sorted ascending order according to row sum)\n",
    "mtx = mtx.iloc[mtx.sum(axis=1).argsort()]\n",
    "mtx = mtx[~mtx.index.duplicated(keep='last')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(57820, 121)"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mtx.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Gene conversion attempt I: Using ensembl conversion tool (https://www.ensembl.org/biomart/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "ensg_hgnc = pd.read_csv('/Users/bramvandesande/Projects/lcb/resources/ensg2hgnc.tsv', sep='\\t', index_col=0).dropna()\n",
    "ensg2hgnc = ensg_hgnc.head().to_dict()['HGNC symbol']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ENSG00000209082',\n",
       " 'ENSG00000210049',\n",
       " 'ENSG00000210077',\n",
       " 'ENSG00000210082',\n",
       " 'ENSG00000211459'}"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(mtx.index).intersection(set(ensg2hgnc.keys()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Gene conversion attempt II: via mygene REST API "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import mygene\n",
    "mg = mygene.MyGeneInfo()\n",
    "out = mg.querymany(list(mtx.index.values), scopes='ensemblgene', fields='ensemblgene,symbol,name', species='human')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "ensg2symbol = dict(zip(map(lambda d: d.get('_id', '?'), out), \n",
    "                       map(lambda d: d.get('symbol', '?'), out)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "del ensg2symbol['?']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "52739"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ensg2symbol)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx.index = list(map(lambda g: ensg2symbol.get(g, '?'), mtx.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM1887287_PDX_pRCC_SC_72</th>\n",
       "      <th>GSM1887294_PDX_pRCC_SC_88</th>\n",
       "      <th>GSM1887230_PDX_mRCC_SC_52</th>\n",
       "      <th>GSM1887285_PDX_pRCC_SC_68</th>\n",
       "      <th>GSM1887254_PDX_pRCC_SC_05</th>\n",
       "      <th>GSM1887274_PDX_pRCC_SC_55</th>\n",
       "      <th>GSM1887335_Pt_mRCC_SC_57</th>\n",
       "      <th>GSM1887336_Pt_mRCC_SC_58</th>\n",
       "      <th>GSM1887286_PDX_pRCC_SC_70</th>\n",
       "      <th>GSM1887345_Pt_mRCC_SC_84</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1887225_PDX_mRCC_SC_20</th>\n",
       "      <th>GSM1887218_PDX_mRCC_SC_04</th>\n",
       "      <th>GSM1887321_Pt_mRCC_SC_21</th>\n",
       "      <th>GSM1887261_PDX_pRCC_SC_27</th>\n",
       "      <th>GSM1887277_PDX_pRCC_SC_58</th>\n",
       "      <th>GSM1887223_PDX_mRCC_SC_14</th>\n",
       "      <th>GSM1887312_Pt_mRCC_SC_10</th>\n",
       "      <th>GSM1887340_Pt_mRCC_SC_69</th>\n",
       "      <th>GSM1887244_PDX_mRCC_SC_87</th>\n",
       "      <th>GSM1887264_PDX_pRCC_SC_34</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HMGB3P8</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FTLP1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>?</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>?</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL360089.1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 121 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            GSM1887287_PDX_pRCC_SC_72  GSM1887294_PDX_pRCC_SC_88  \\\n",
       "HMGB3P8                           0.0                        0.0   \n",
       "FTLP1                             0.0                        0.0   \n",
       "?                                 0.0                        0.0   \n",
       "?                                 0.0                        0.0   \n",
       "AL360089.1                        0.0                        0.0   \n",
       "\n",
       "            GSM1887230_PDX_mRCC_SC_52  GSM1887285_PDX_pRCC_SC_68  \\\n",
       "HMGB3P8                           0.0                        0.0   \n",
       "FTLP1                             0.0                        0.0   \n",
       "?                                 0.0                        0.0   \n",
       "?                                 0.0                        0.0   \n",
       "AL360089.1                        0.0                        0.0   \n",
       "\n",
       "            GSM1887254_PDX_pRCC_SC_05  GSM1887274_PDX_pRCC_SC_55  \\\n",
       "HMGB3P8                           0.0                        0.0   \n",
       "FTLP1                             0.0                        0.0   \n",
       "?                                 0.0                        0.0   \n",
       "?                                 0.0                        0.0   \n",
       "AL360089.1                        0.0                        0.0   \n",
       "\n",
       "            GSM1887335_Pt_mRCC_SC_57  GSM1887336_Pt_mRCC_SC_58  \\\n",
       "HMGB3P8                          0.0                       0.0   \n",
       "FTLP1                            0.0                       0.0   \n",
       "?                                0.0                       0.0   \n",
       "?                                0.0                       0.0   \n",
       "AL360089.1                       0.0                       0.0   \n",
       "\n",
       "            GSM1887286_PDX_pRCC_SC_70  GSM1887345_Pt_mRCC_SC_84  \\\n",
       "HMGB3P8                           0.0                       0.0   \n",
       "FTLP1                             0.0                       0.0   \n",
       "?                                 0.0                       0.0   \n",
       "?                                 0.0                       0.0   \n",
       "AL360089.1                        0.0                       0.0   \n",
       "\n",
       "                      ...              GSM1887225_PDX_mRCC_SC_20  \\\n",
       "HMGB3P8               ...                                    0.0   \n",
       "FTLP1                 ...                                    0.0   \n",
       "?                     ...                                    0.0   \n",
       "?                     ...                                    0.0   \n",
       "AL360089.1            ...                                    0.0   \n",
       "\n",
       "            GSM1887218_PDX_mRCC_SC_04  GSM1887321_Pt_mRCC_SC_21  \\\n",
       "HMGB3P8                           0.0                       0.0   \n",
       "FTLP1                             0.0                       0.0   \n",
       "?                                 0.0                       0.0   \n",
       "?                                 0.0                       0.0   \n",
       "AL360089.1                        0.0                       0.0   \n",
       "\n",
       "            GSM1887261_PDX_pRCC_SC_27  GSM1887277_PDX_pRCC_SC_58  \\\n",
       "HMGB3P8                           0.0                        0.0   \n",
       "FTLP1                             0.0                        0.0   \n",
       "?                                 0.0                        0.0   \n",
       "?                                 0.0                        0.0   \n",
       "AL360089.1                        0.0                        0.0   \n",
       "\n",
       "            GSM1887223_PDX_mRCC_SC_14  GSM1887312_Pt_mRCC_SC_10  \\\n",
       "HMGB3P8                           0.0                       0.0   \n",
       "FTLP1                             0.0                       0.0   \n",
       "?                                 0.0                       0.0   \n",
       "?                                 0.0                       0.0   \n",
       "AL360089.1                        0.0                       0.0   \n",
       "\n",
       "            GSM1887340_Pt_mRCC_SC_69  GSM1887244_PDX_mRCC_SC_87  \\\n",
       "HMGB3P8                          0.0                        0.0   \n",
       "FTLP1                            0.0                        0.0   \n",
       "?                                0.0                        0.0   \n",
       "?                                0.0                        0.0   \n",
       "AL360089.1                       0.0                        0.0   \n",
       "\n",
       "            GSM1887264_PDX_pRCC_SC_34  \n",
       "HMGB3P8                           0.0  \n",
       "FTLP1                             0.0  \n",
       "?                                 0.0  \n",
       "?                                 0.0  \n",
       "AL360089.1                        0.0  \n",
       "\n",
       "[5 rows x 121 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mtx.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx = mtx[mtx.index != '?']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(28665, 121)"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mtx.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx = mtx.iloc[mtx.sum(axis=1).argsort()]\n",
    "mtx = mtx[~mtx.index.duplicated(keep='last')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(27530, 121)"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mtx.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ABCC6P1',\n",
       " 'ACTR3BP2',\n",
       " 'AGAP11',\n",
       " 'AKR7A2P1',\n",
       " 'ANKRD26P1',\n",
       " 'ANKRD30BP2',\n",
       " 'ANKRD36',\n",
       " 'ANP32C',\n",
       " 'ANXA2P1',\n",
       " 'ANXA2P2',\n",
       " 'ANXA2P3',\n",
       " 'AOX2P',\n",
       " 'AQP7P1',\n",
       " 'ARGFXP2',\n",
       " 'ASB18',\n",
       " 'ASB9P1',\n",
       " 'ATP5EP2',\n",
       " 'ATXN8OS',\n",
       " 'AURKAPS1',\n",
       " 'BAGE2',\n",
       " 'BCRP2',\n",
       " 'BCRP3',\n",
       " 'BMS1P4',\n",
       " 'BTF3P11',\n",
       " 'C10orf111',\n",
       " 'C10orf25',\n",
       " 'C10orf91',\n",
       " 'C12orf77',\n",
       " 'C15orf32',\n",
       " 'C15orf56',\n",
       " 'C17orf102',\n",
       " 'C17orf77',\n",
       " 'C17orf82',\n",
       " 'C1orf229',\n",
       " 'C20orf197',\n",
       " 'C20orf203',\n",
       " 'C22orf24',\n",
       " 'C2orf27A',\n",
       " 'C2orf27B',\n",
       " 'C2orf48',\n",
       " 'C3orf70',\n",
       " 'C3orf79',\n",
       " 'C6orf99',\n",
       " 'C7orf65',\n",
       " 'C7orf66',\n",
       " 'C7orf69',\n",
       " 'C7orf71',\n",
       " 'C9orf106',\n",
       " 'C9orf139',\n",
       " 'C9orf163',\n",
       " 'C9orf170',\n",
       " 'CCL27',\n",
       " 'CCT6P1',\n",
       " 'CDC14C',\n",
       " 'CEACAM18',\n",
       " 'CELP',\n",
       " 'CEP170P1',\n",
       " 'CES1P2',\n",
       " 'CETN4P',\n",
       " 'CHKB-CPT1B',\n",
       " 'CIDECP',\n",
       " 'CLEC4GP1',\n",
       " 'CLRN3',\n",
       " 'COG8',\n",
       " 'COL6A4P2',\n",
       " 'CRX',\n",
       " 'CSNK1A1P1',\n",
       " 'CSPG4P2Y',\n",
       " 'CTAGE10P',\n",
       " 'CTAGE11P',\n",
       " 'CXADRP2',\n",
       " 'CXADRP3',\n",
       " 'CXCR2P1',\n",
       " 'CYP2S1',\n",
       " 'CYP4F24P',\n",
       " 'DDX11L2',\n",
       " 'DEFT1P2',\n",
       " 'DGCR10',\n",
       " 'DGCR5',\n",
       " 'DHRS4L1',\n",
       " 'DHX40P1',\n",
       " 'DIO3OS',\n",
       " 'DIRC3',\n",
       " 'DLEU2',\n",
       " 'DLEU2L',\n",
       " 'DNAJB3',\n",
       " 'DNM1P35',\n",
       " 'DOC2GP',\n",
       " 'DPH3P1',\n",
       " 'DPRXP4',\n",
       " 'DPY19L1P1',\n",
       " 'DPY19L2P1',\n",
       " 'DSCR4',\n",
       " 'DUX4L4',\n",
       " 'EEF1A1P9',\n",
       " 'EEF1DP3',\n",
       " 'EGFEM1P',\n",
       " 'EP400NL',\n",
       " 'FABP5P3',\n",
       " 'FAM157A',\n",
       " 'FAM157B',\n",
       " 'FAM182B',\n",
       " 'FAM201A',\n",
       " 'FAM41AY1',\n",
       " 'FAM41AY2',\n",
       " 'FAM66A',\n",
       " 'FAM66B',\n",
       " 'FAM66C',\n",
       " 'FAM66D',\n",
       " 'FBXL21',\n",
       " 'FCGR2C',\n",
       " 'FCN1',\n",
       " 'FER1L4',\n",
       " 'FHAD1',\n",
       " 'FKBP1AP1',\n",
       " 'FOLH1B',\n",
       " 'FOXO3B',\n",
       " 'FTH1P3',\n",
       " 'FUNDC2P2',\n",
       " 'GABARAPL3',\n",
       " 'GAGE12B',\n",
       " 'GDF3',\n",
       " 'GEMIN8P4',\n",
       " 'GGT3P',\n",
       " 'GGT8P',\n",
       " 'GK3P',\n",
       " 'GNAZ',\n",
       " 'GOLGA2P3Y',\n",
       " 'GPR21',\n",
       " 'GPR37L1',\n",
       " 'GSTA7P',\n",
       " 'GSTM2P1',\n",
       " 'GTF2H2B',\n",
       " 'GTF2IRD2P1',\n",
       " 'GUCY2GP',\n",
       " 'GUSBP3',\n",
       " 'GUSBP4',\n",
       " 'GYG2P1',\n",
       " 'HAR1A',\n",
       " 'HBBP1',\n",
       " 'HCG18',\n",
       " 'HCG4',\n",
       " 'HERC2P4',\n",
       " 'HERC2P7',\n",
       " 'HILS1',\n",
       " 'HIST2H2BA',\n",
       " 'HLA-H',\n",
       " 'HLA-J',\n",
       " 'HMGB3P1',\n",
       " 'HNRNPA3P1',\n",
       " 'HSD3BP4',\n",
       " 'HSP90AA4P',\n",
       " 'HSP90AA6P',\n",
       " 'HSP90AB2P',\n",
       " 'HSP90AB3P',\n",
       " 'HSP90B3P',\n",
       " 'HSPA7',\n",
       " 'HULC',\n",
       " 'HYALP1',\n",
       " 'IFITM4P',\n",
       " 'IFNA22P',\n",
       " 'IGLL3P',\n",
       " 'ITGB1BP2',\n",
       " 'KCNA6',\n",
       " 'KIR3DP1',\n",
       " 'KIR3DX1',\n",
       " 'KRT16P2',\n",
       " 'KRTAP10-2',\n",
       " 'LCN2',\n",
       " 'LRRC37BP1',\n",
       " 'LY6G6E',\n",
       " 'LYPLA2P1',\n",
       " 'MAGEA5',\n",
       " 'MEG8',\n",
       " 'MEIS3P1',\n",
       " 'MESTIT1',\n",
       " 'MIR130B',\n",
       " 'MIR548C',\n",
       " 'MKRN9P',\n",
       " 'MMP23A',\n",
       " 'MORF4',\n",
       " 'MOXD2P',\n",
       " 'MRPL45P2',\n",
       " 'MSLNL',\n",
       " 'MST1P2',\n",
       " 'MSTO2P',\n",
       " 'MSX2P1',\n",
       " 'MT1DP',\n",
       " 'MTMR9LP',\n",
       " 'MTRNR2L2',\n",
       " 'MTRNR2L9',\n",
       " 'MUC2',\n",
       " 'MYH16',\n",
       " 'MYO15B',\n",
       " 'NACAP1',\n",
       " 'NBPF22P',\n",
       " 'NBPF7',\n",
       " 'NF1P1',\n",
       " 'NKAIN3',\n",
       " 'NRADDP',\n",
       " 'NSFP1',\n",
       " 'NUDT9P1',\n",
       " 'OGFR',\n",
       " 'OR2A9P',\n",
       " 'OR2M1P',\n",
       " 'OR2W5',\n",
       " 'OR4K15',\n",
       " 'OR5E1P',\n",
       " 'OR6K6',\n",
       " 'OR6W1P',\n",
       " 'OR7E37P',\n",
       " 'OR7E5P',\n",
       " 'OR9K2',\n",
       " 'OVCH1',\n",
       " 'P2RX6P',\n",
       " 'PA2G4P4',\n",
       " 'PABPC1P2',\n",
       " 'PCDHB19P',\n",
       " 'PCDHB3',\n",
       " 'PCNAP1',\n",
       " 'PDXDC2P',\n",
       " 'PDZK1P1',\n",
       " 'PI4KAP2',\n",
       " 'PIP5K1P1',\n",
       " 'PIPSL',\n",
       " 'PKD1P1',\n",
       " 'PLGLA',\n",
       " 'PMCHL1',\n",
       " 'PMS2CL',\n",
       " 'PMS2P1',\n",
       " 'PMS2P11',\n",
       " 'PMS2P4',\n",
       " 'PMS2P5',\n",
       " 'POLR2A',\n",
       " 'POLR2J4',\n",
       " 'POM121L10P',\n",
       " 'POM121L1P',\n",
       " 'POM121L4P',\n",
       " 'POM121L8P',\n",
       " 'POTEKP',\n",
       " 'PPDPF',\n",
       " 'PPIEL',\n",
       " 'PPP1R2P1',\n",
       " 'PPP4R1L',\n",
       " 'PRNT',\n",
       " 'PROKR1',\n",
       " 'PRR21',\n",
       " 'PRSS50',\n",
       " 'PTTG3P',\n",
       " 'PVT1',\n",
       " 'PWRN2',\n",
       " 'RACGAP1P',\n",
       " 'RBMY1A3P',\n",
       " 'RBMY3AP',\n",
       " 'RFPL1S',\n",
       " 'RFPL3S',\n",
       " 'RHOXF2B',\n",
       " 'RNF138P1',\n",
       " 'RNF5P1',\n",
       " 'RPL13P5',\n",
       " 'RPL18AP3',\n",
       " 'RPL21P28',\n",
       " 'RPL21P44',\n",
       " 'RPL23AP32',\n",
       " 'RPL23AP53',\n",
       " 'RPL23AP64',\n",
       " 'RPL23AP7',\n",
       " 'RPL23AP82',\n",
       " 'RPL23P8',\n",
       " 'RPL32P3',\n",
       " 'RPL5',\n",
       " 'RPS10P7',\n",
       " 'RPS15AP10',\n",
       " 'RPS26P11',\n",
       " 'RPS27',\n",
       " 'RPS2P32',\n",
       " 'RPS7P5',\n",
       " 'RPSAP58',\n",
       " 'RPSAP9',\n",
       " 'RRN3P1',\n",
       " 'RRN3P2',\n",
       " 'RRN3P3',\n",
       " 'RYBP',\n",
       " 'SAMD1',\n",
       " 'SBDSP1',\n",
       " 'SCARNA11',\n",
       " 'SCARNA15',\n",
       " 'SCARNA16',\n",
       " 'SCARNA17',\n",
       " 'SCARNA18',\n",
       " 'SCARNA20',\n",
       " 'SCARNA21',\n",
       " 'SCARNA6',\n",
       " 'SDC4P',\n",
       " 'SDHAP2',\n",
       " 'SDHAP3',\n",
       " 'SEPT7P2',\n",
       " 'SGCZ',\n",
       " 'SH3GL1P1',\n",
       " 'SH3GL1P2',\n",
       " 'SIGLEC16',\n",
       " 'SLC35E3',\n",
       " 'SLC7A5P1',\n",
       " 'SLC7A5P2',\n",
       " 'SNHG1',\n",
       " 'SNHG10',\n",
       " 'SNHG5',\n",
       " 'SNHG6',\n",
       " 'SNHG9',\n",
       " 'SNORA1',\n",
       " 'SNORA11',\n",
       " 'SNORA12',\n",
       " 'SNORA15',\n",
       " 'SNORA17',\n",
       " 'SNORA18',\n",
       " 'SNORA19',\n",
       " 'SNORA20',\n",
       " 'SNORA21',\n",
       " 'SNORA22',\n",
       " 'SNORA24',\n",
       " 'SNORA25',\n",
       " 'SNORA26',\n",
       " 'SNORA27',\n",
       " 'SNORA3',\n",
       " 'SNORA30',\n",
       " 'SNORA31',\n",
       " 'SNORA32',\n",
       " 'SNORA33',\n",
       " 'SNORA35',\n",
       " 'SNORA38',\n",
       " 'SNORA4',\n",
       " 'SNORA40',\n",
       " 'SNORA42',\n",
       " 'SNORA43',\n",
       " 'SNORA44',\n",
       " 'SNORA46',\n",
       " 'SNORA48',\n",
       " 'SNORA51',\n",
       " 'SNORA57',\n",
       " 'SNORA58',\n",
       " 'SNORA62',\n",
       " 'SNORA63',\n",
       " 'SNORA64',\n",
       " 'SNORA67',\n",
       " 'SNORA68',\n",
       " 'SNORA69',\n",
       " 'SNORA70',\n",
       " 'SNORA71B',\n",
       " 'SNORA72',\n",
       " 'SNORA75',\n",
       " 'SNORA76',\n",
       " 'SNORA77',\n",
       " 'SNORA8',\n",
       " 'SNORA81',\n",
       " 'SNORA9',\n",
       " 'SNORD10',\n",
       " 'SNORD104',\n",
       " 'SNORD105B',\n",
       " 'SNORD11',\n",
       " 'SNORD111',\n",
       " 'SNORD111B',\n",
       " 'SNORD114-4',\n",
       " 'SNORD115-24',\n",
       " 'SNORD116-17',\n",
       " 'SNORD116-19',\n",
       " 'SNORD116-26',\n",
       " 'SNORD116-27',\n",
       " 'SNORD11B',\n",
       " 'SNORD123',\n",
       " 'SNORD126',\n",
       " 'SNORD127',\n",
       " 'SNORD12B',\n",
       " 'SNORD12C',\n",
       " 'SNORD16',\n",
       " 'SNORD19',\n",
       " 'SNORD19B',\n",
       " 'SNORD2',\n",
       " 'SNORD23',\n",
       " 'SNORD27',\n",
       " 'SNORD29',\n",
       " 'SNORD30',\n",
       " 'SNORD33',\n",
       " 'SNORD35B',\n",
       " 'SNORD36B',\n",
       " 'SNORD37',\n",
       " 'SNORD38B',\n",
       " 'SNORD41',\n",
       " 'SNORD44',\n",
       " 'SNORD46',\n",
       " 'SNORD5',\n",
       " 'SNORD51',\n",
       " 'SNORD56',\n",
       " 'SNORD58C',\n",
       " 'SNORD60',\n",
       " 'SNORD63',\n",
       " 'SNORD64',\n",
       " 'SNORD65',\n",
       " 'SNORD66',\n",
       " 'SNORD67',\n",
       " 'SNORD68',\n",
       " 'SNORD70',\n",
       " 'SNORD74',\n",
       " 'SNORD75',\n",
       " 'SNORD77',\n",
       " 'SNORD78',\n",
       " 'SNORD79',\n",
       " 'SNORD81',\n",
       " 'SNORD87',\n",
       " 'SNORD88B',\n",
       " 'SNORD88C',\n",
       " 'SNORD95',\n",
       " 'SNRPA1',\n",
       " 'SNRPD2P2',\n",
       " 'SPINT4',\n",
       " 'SPRR2C',\n",
       " 'SRD5A1',\n",
       " 'SRD5A1P1',\n",
       " 'SSX6',\n",
       " 'SSX8',\n",
       " 'ST13P4',\n",
       " 'STAG3L1',\n",
       " 'SUGT1P1',\n",
       " 'SUMO1P1',\n",
       " 'SUMO1P3',\n",
       " 'SYS1-DBNDD2',\n",
       " 'TBC1D3P2',\n",
       " 'TERC',\n",
       " 'TEX21P',\n",
       " 'TFAMP1',\n",
       " 'TIGD1',\n",
       " 'TIMM22',\n",
       " 'TM4SF20',\n",
       " 'TMED10P1',\n",
       " 'TMEM133',\n",
       " 'TMPRSS11BNL',\n",
       " 'TNXA',\n",
       " 'TOP1P1',\n",
       " 'TP53TG1',\n",
       " 'TPI1P3',\n",
       " 'TPTE2P1',\n",
       " 'TPTE2P2',\n",
       " 'TPTE2P3',\n",
       " 'TSNAX-DISC1',\n",
       " 'TSPY26P',\n",
       " 'TTC24',\n",
       " 'TTC3P1',\n",
       " 'TTLL8',\n",
       " 'TTTY14',\n",
       " 'TTTY2',\n",
       " 'TTTY2B',\n",
       " 'TUBBP5',\n",
       " 'TUG1',\n",
       " 'TYRO3P',\n",
       " 'UBE2MP1',\n",
       " 'UOX',\n",
       " 'VENTXP7',\n",
       " 'WASH2P',\n",
       " 'WASH3P',\n",
       " 'WDR49',\n",
       " 'WT1-AS',\n",
       " 'XKRY',\n",
       " 'XKRY2',\n",
       " 'ZDHHC8P1',\n",
       " 'ZFP91-CNTF',\n",
       " 'ZNF137P',\n",
       " 'ZNF204P',\n",
       " 'ZNF300P1',\n",
       " 'ZNF658B',\n",
       " 'ZNF826P',\n",
       " 'ZNF833P',\n",
       " 'ZNF847P',\n",
       " 'ZNF890P',\n",
       " 'ZSCAN12P1']"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(mtx[mtx.index.isin(genes_in_db)].index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "473"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(mtx[mtx.index.isin(genes_in_db)].index)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.6.2 (pyscenic_dev)",
   "language": "python",
   "name": "pyscenic_dev"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
